# NER tagging

# Data download and description

In [ ]:
from urllib.request import urlretrieve
urlretrieve('https://raw.githubusercontent.com/pranabsarkar/Conll_task/master/conll-2003/eng.train','eng.train')
urlretrieve('https://raw.githubusercontent.com/pranabsarkar/Conll_task/master/conll-2003/eng.testa','eng.testa')

istream = open('eng.train')
for idx, line in enumerate(istream):
  print(line.strip())
  if idx >=20:
    break
istream.close()


-DOCSTART- -X- -X- O

EU NNP I-NP I-ORG
rejects VBZ I-VP O
German JJ I-NP I-MISC
call NN I-NP O
to TO I-VP O
boycott VB I-VP O
British JJ I-NP I-MISC
lamb NN I-NP O
. . O O

Peter NNP I-NP I-PER
Blackburn NNP I-NP I-PER

BRUSSELS NNP I-NP I-LOC
1996-08-22 CD I-NP O

The DT I-NP O
European NNP I-NP I-ORG
Commission NNP I-NP I-ORG


The CONLL 2003 dataset encodes each token on a single line followed by its annotation:

> (token,tag,chunk,named entity)

The NER tags follow the IOB convention:
* **I** means **Inside** (part of a named entity);
* **B** means **Begin** (starting a new entity);
* **O** means **Outside** (not part of a named entity).

The **I** and **B** Tag are followed by a specifier (for example, PER=person, ORG=Organisation).

# Data preprocessing

In [ ]:
def vocabulary(filename, input_vocab, padding='<pad>', unknown='<unk>'):
    # input_vocab is a boolean flag to tell if we extract input or output vocabulary

    idx2sym = {}
    sym2idx = {}

    cur_idx = 0
    # Add pad and unk tokens to the vocab
    if padding:
      idx2sym[cur_idx] = padding
      sym2idx[padding] = cur_idx
      cur_idx += 1
    if unknown:
      idx2sym[cur_idx] = unknown
      sym2idx[unknown] = cur_idx
      cur_idx += 1

    with open(filename, 'r', encoding='utf-8') as file:
      for line in file:
        line = line.strip()

        if not line or line.startswith('-DOCSTART-'):
          continue  # Skip empty lines and metadata

        parts = line.split()
        if input_vocab:
          token = parts[0]  # The token is the first column
          if token not in sym2idx:
            idx2sym[cur_idx] = token
            sym2idx[token] = cur_idx
            cur_idx += 1
        else:
          tag = parts[-1]  # The NER tag is the last column
          if tag not in sym2idx:
            idx2sym[cur_idx] = tag
            sym2idx[tag] = cur_idx
            cur_idx += 1

    return idx2sym, sym2idx

In [ ]:
def pad_sequence(sequence, pad_size, pad_token):
    # returns a list with pad tokens
    if sequence is None:
      print("None sequence is found and replaced with pad tokens.")
      return [pad_token] * pad_size
    return sequence[:pad_size] + [pad_token] * (pad_size - len(sequence))

def code_sequence(sequence, coding_map, unk_token=None):
    # takes a list of strs and returns a list of ints
    if sequence is None:
      return []
    unk_index = coding_map.get(unk_token, None)
    return [coding_map.get(token, unk_index) for token in sequence]

def decode_sequence(sequence, decoding_map):
    # takes a list of ints and returns a list of strs
    return [decoding_map.get(index, '<unk>') for index in sequence]

In [ ]:
# Testing
# Sample vocab maps
sym2idx = {'EU': 1, 'NNP': 2, 'I-NP': 3, 'I-ORG': 4, '<pad>': 0, '<unk>': 5}
idx2sym = {v: k for k, v in sym2idx.items()}

sequence = ['EU', 'NNP', 'I-NP', 'I-ORG']
pad_size = 6
padded_seq= pad_sequence(sequence, pad_size, pad_token='<pad>')
print(padded_seq)

encoded_seq = code_sequence(padded_seq, sym2idx, unk_token='<unk>')
print(encoded_seq)

# Test decoding
decoded_seq = decode_sequence(encoded_seq, idx2sym)
print(decoded_seq)


['EU', 'NNP', 'I-NP', 'I-ORG', '<pad>', '<pad>']
[1, 2, 3, 4, 0, 0]
['EU', 'NNP', 'I-NP', 'I-ORG', '<pad>', '<pad>']


# Data generator

In [ ]:
def read_conll_tokens(conllfilename):
    # reads a CONLL 2003 file and returns a list of sentences.
    # A sentence is a list of strings (tokens)
    sentences = []
    cur_sent = []

    with open(conllfilename, 'r') as f:
      for line in f:
        line = line.strip()
        if not line:
          if cur_sent:
            sentences.append(cur_sent)
            cur_sent = []
        elif not line.startswith('-DOCSTART-'):
          parts = line.split()
          cur_sent.append(parts[0])   # token

    # Add the last sentence if the file does not end with a newline
    if cur_sent:
      sentences.append(cur_sent)

    return sentences

def read_conll_tags(conllfilename):
    # reads a CONLL 2003 file and returns a list of sentences.
    # A sentence is a list of strings (tokens)
    sentences = []
    cur_sent = []

    with open(conllfilename, 'r') as f:
      for line in f:
        line = line.strip()
        if not line:
          if cur_sent:
            sentences.append(cur_sent)
            cur_sent = []
        elif not line.startswith('-DOCSTART-'):
          parts = line.split()
          cur_sent.append(parts[-1])   # NER tag

    if cur_sent:
      sentences.append(cur_sent)

    return sentences

In [ ]:
# Test on first sentence
sample_tokens = tokens[0]
sample_tags = tags[0]
print(sample_tokens)
print(sample_tags)
print()

# Pad sequences
padded_tokens = pad_sequence(sample_tokens, pad_size=12, pad_token='<pad>')
padded_tags = pad_sequence(sample_tags, pad_size=12, pad_token='<pad>')
print(padded_tokens)
print(padded_tags)
print()

# Encode sequences
encoded_tokens = code_sequence(padded_tokens, input_sym2idx, unk_token='<unk>')
encoded_tags = code_sequence(padded_tags, output_sym2idx, unk_token='<unk>')
print(encoded_tokens)
print(encoded_tags)
print()

# Decode sequences
decoded_tokens = decode_sequence(encoded_tokens, input_idx2sym)
decoded_tags = decode_sequence(encoded_tags, output_idx2sym)
print(decoded_tokens)
print(decoded_tags)

['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.']
['I-ORG', 'O', 'I-MISC', 'O', 'O', 'O', 'I-MISC', 'O', 'O']

['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.', '<pad>', '<pad>', '<pad>']
['I-ORG', 'O', 'I-MISC', 'O', 'O', 'O', 'I-MISC', 'O', 'O', '<pad>', '<pad>', '<pad>']

[2, 3, 4, 5, 6, 7, 8, 9, 10, 0, 0, 0]
[2, 3, 4, 3, 3, 3, 4, 3, 3, 0, 0, 0]

['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.', '<pad>', '<pad>', '<pad>']
['I-ORG', 'O', 'I-MISC', 'O', 'O', 'O', 'I-MISC', 'O', 'O', '<pad>', '<pad>', '<pad>']




Now we implement the class. You will rely on the helper functions designed above in order to fill in the blanks in the constructor.

In [ ]:
import torch
import torch.nn as nn
from random import shuffle

class DataGenerator:

        def __init__(self,conllfilename, parentgenerator = None, pad_token='<pad>',unk_token='<unk>'):

              if parentgenerator is not None: 
                  self.pad_token = parentgenerator.pad_token
                  self.unk_token = parentgenerator.unk_token
                  self.input_sym2idx = parentgenerator.input_sym2idx
                  self.input_idx2sym = parentgenerator.input_idx2sym
                  self.output_sym2idx = parentgenerator.output_sym2idx
                  self.output_idx2sym = parentgenerator.output_idx2sym
              else:                           # Creates new encodings
                  self.pad_token = pad_token
                  self.unk_token = unk_token
                  # Creates 4 encoding maps from datafile
                  self.input_idx2sym,self.input_sym2idx = vocabulary(conllfilename,input_vocab=True,padding=pad_token,unknown=unk_token)
                  self.output_idx2sym,self.output_sym2idx = vocabulary(conllfilename,input_vocab=False,padding=pad_token,unknown=unk_token)

              # store the conll dataset with sentence structure (a list of lists of strings) in the following fields
              self.Xtokens = read_conll_tokens(conllfilename)
              self.Ytokens = read_conll_tags(conllfilename)

        def generate_batches(self,batch_size):

              # Generator function yielding one batch after another. Batches are lists of lists

              assert(len(self.Xtokens) == len(self.Ytokens))

              N = len(self.Xtokens)
              idxes = list(range(N))

              # Data ordering
              shuffle(idxes)
              idxes.sort(key=lambda idx: len(self.Xtokens[idx]))

              # batch generation
              bstart = 0
              while bstart < N:
                 bend = min(bstart+batch_size,N)
                 batch_idxes = idxes[bstart:bend]
                 batch_len = max(len(self.Xtokens[idx]) for idx in batch_idxes)

                 seqX = [pad_sequence(self.Xtokens[idx],batch_len,self.pad_token) for idx in batch_idxes]
                 seqY = [pad_sequence(self.Ytokens[idx],batch_len,self.pad_token) for idx in batch_idxes]
                 seqX = [code_sequence(seq,self.input_sym2idx,self.unk_token) for seq in seqX]
                 seqY = [code_sequence(seq,self.output_sym2idx) for seq in seqY]
                 #print(seqX, seqY)

                 assert(len(seqX) == len(seqY))
                 yield (seqX,seqY)
                 bstart += batch_size

Implementation of the tagger.

* Implemention of parameter allocation (the embedding layer, the LSTM (or bi-LSTM) layer and the Linear Layer);
* Implemention of the forward method;
* Implemention of the train method.

In [ ]:
import torch.optim as optim

class NERtagger(nn.Module):

      def __init__(self,traingenerator, embedding_size,hidden_size,device='cuda'):
        super(NERtagger, self).__init__()
        self.embedding_size = embedding_size
        self.hidden_size = hidden_size
        self.allocate_params(traingenerator,device)

      def load(self,filename):
        self.load_state_dict(torch.load(filename))

      def allocate_params(self,datagenerator,device):
        # Get vocabulary and number of output classes
        vocab_size = len(datagenerator.input_sym2idx)  # Size of the input vocabulary
        num_classes = len(datagenerator.output_sym2idx)  # Number of output tags

        # Embedding layer
        self.embedding = nn.Embedding(vocab_size, self.embedding_size, padding_idx=datagenerator.input_sym2idx[datagenerator.pad_token]).to(device)

        # Bi-LSTM layer
        self.lstm = nn.LSTM(
          input_size=self.embedding_size,
          hidden_size=self.hidden_size,
          num_layers=1,   # by default
          batch_first=True,
          bidirectional=True  # Enables Bi-LSTM
       ).to(device)

        # Linear layer
        self.linear = nn.Linear(self.hidden_size * 2, num_classes).to(device)  # Bi-LSTM output is 2 * hidden_size

      def forward(self,Xinput):
        # prediction steps
        # Embedding layer
        embeddings = self.embedding(Xinput)  # Shape: (batch_size, seq_len, embedding_size)

        # LSTM layer
        lstm_out, _ = self.lstm(embeddings)  # Shape: (batch_size, seq_len, hidden_size * 2)

        # Linear layer
        logits = self.linear(lstm_out)  # Shape: (batch_size, seq_len, num_classes)

        return logits

      def train_model(self,traingenerator,validgenerator,epochs,batch_size,device='cuda',learning_rate=0.001):

        self.minloss = 10000000 # the min loss found so far on validation data

        self.to(device)
        optimizer = optim.Adam(self.parameters(), lr=learning_rate)

        pad_index = traingenerator.output_sym2idx[traingenerator.pad_token]
        loss_fnc = nn.CrossEntropyLoss(ignore_index=pad_index)

        for epoch in range(1, epochs + 1):
          print(f"\nEpoch {epoch}/{epochs}")
          self.train()  # Set model to training mode

          batch_losses = []
          batch_accuracies = []

          for seqX, seqY in traingenerator.generate_batches(batch_size):
            X = torch.LongTensor(seqX).to(device)
            Y = torch.LongTensor(seqY).to(device)

            Yhat = self.forward(X)

            # Flatten outputs for loss computation
            batch_size, seq_len = Y.shape
            Yhat = Yhat.view(batch_size * seq_len, -1)
            Y = Y.view(batch_size * seq_len)

            # Loss computation
            loss = loss_fnc(Yhat, Y)
            batch_losses.append(loss.item())

            # Backward pass and optimization
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            # Accuracy computation
            mask = (Y != pad_index)
            Yargmax = torch.argmax(Yhat, dim=1)
            correct = torch.sum((Yargmax == Y) * mask)
            total = torch.sum(mask)
            batch_accuracies.append(float(correct) / float(total))

          # Epoch summary
          train_loss = sum(batch_losses) / len(batch_losses)
          train_accuracy = sum(batch_accuracies) / len(batch_accuracies)
          print(f"[train] Epoch {epoch} mean loss = {train_loss:.4f} | mean accuracy = {train_accuracy:.4f}")

          # Validate model
          valid_loss, valid_accuracy = self.validate(validgenerator, batch_size, device, save_min_model=True)
          print(f"[valid] Epoch {epoch} mean loss = {valid_loss:.4f} | mean accuracy = {valid_accuracy:.4f}")

      def validate(self,datagenerator,batch_size,device='cuda',save_min_model=False):

          batch_accurracies = []
          batch_losses = []

          device = torch.device(device)
          pad_index = datagenerator.output_sym2idx[datagenerator.pad_token]
          loss_fnc = nn.CrossEntropyLoss(ignore_index=pad_index)

          for (seqX,seqY) in datagenerator.generate_batches(batch_size):
                with torch.no_grad():
                  X = torch.LongTensor(seqX).to(device)
                  Y = torch.LongTensor(seqY).to(device)

                  Yhat = self.forward(X)

                  #Flattening and loss computation
                  batch_size,seq_len = Y.shape
                  Yhat = Yhat.view(batch_size*seq_len,-1)
                  Y = Y.view(batch_size*seq_len)
                  loss = loss_fnc(Yhat,Y)
                  batch_losses.append(loss.item())

                  #Accurracy computation
                  mask = (Y != pad_index)
                  Yargmax = torch.argmax(Yhat,dim=1)
                  correct = torch.sum((Yargmax == Y) * mask)
                  total = torch.sum(mask)
                  batch_accurracies.append(float(correct)/float(total))

          L = len(batch_losses)
          valid_loss = sum(batch_losses)/L
          valid_accuracy = sum(batch_accurracies) / L

          if save_min_model and valid_loss < self.minloss:
            self.minloss = valid_loss
            torch.save(self.state_dict(), 'tagger_params.pt')

          return valid_loss, valid_accuracy

# Main program

In [ ]:
# Load the full dataset
full_trainset = DataGenerator('eng.train')

# Create a small sample from the full dataset
sample_size = 15  # Choose a small sample size for testing
small_trainset = DataGenerator('eng.train')

# Limit the Xtokens and Ytokens to the first few examples
small_trainset.Xtokens = full_trainset.Xtokens[:sample_size]
small_trainset.Ytokens = full_trainset.Ytokens[:sample_size]

In [ ]:
# Initialize the model
tagger = NERtagger(small_trainset, embedding_size=64, hidden_size=128, device='cuda')

# Train on the small dataset
tagger.train_model(traingenerator=small_trainset, validgenerator=small_trainset, epochs=4, batch_size=4, device='cuda', learning_rate=0.001)



Epoch 1/4
[train] Epoch 1 mean loss = 2.2258 | mean accuracy = 0.4242
[valid] Epoch 1 mean loss = 2.1146 | mean accuracy = 0.7474

Epoch 2/4
[train] Epoch 2 mean loss = 2.0262 | mean accuracy = 0.7904
[valid] Epoch 2 mean loss = 1.8731 | mean accuracy = 0.8416

Epoch 3/4
[train] Epoch 3 mean loss = 1.7318 | mean accuracy = 0.8416
[valid] Epoch 3 mean loss = 1.4944 | mean accuracy = 0.8434

Epoch 4/4
[train] Epoch 4 mean loss = 1.1980 | mean accuracy = 0.8434
[valid] Epoch 4 mean loss = 0.8577 | mean accuracy = 0.8434


In [ ]:
# Train on the full dataset
trainset = DataGenerator('eng.train')
validset = DataGenerator('eng.testa',parentgenerator = trainset)
tagger   = NERtagger(trainset,embedding_size=64,hidden_size=128,device='cuda')
tagger.train_model(traingenerator=trainset,validgenerator=validset,epochs=10,batch_size=32,device='cuda',learning_rate=0.001)


Epoch 1/10
[train] Epoch 1 mean loss = 0.7072 | mean accuracy = 0.7990
[valid] Epoch 1 mean loss = 0.5355 | mean accuracy = 0.8359

Epoch 2/10
[train] Epoch 2 mean loss = 0.3694 | mean accuracy = 0.8854
[valid] Epoch 2 mean loss = 0.3516 | mean accuracy = 0.8934

Epoch 3/10
[train] Epoch 3 mean loss = 0.2167 | mean accuracy = 0.9320
[valid] Epoch 3 mean loss = 0.2610 | mean accuracy = 0.9199

Epoch 4/10
[train] Epoch 4 mean loss = 0.1244 | mean accuracy = 0.9615
[valid] Epoch 4 mean loss = 0.2249 | mean accuracy = 0.9343

Epoch 5/10
[train] Epoch 5 mean loss = 0.0828 | mean accuracy = 0.9741
[valid] Epoch 5 mean loss = 0.2328 | mean accuracy = 0.9338

Epoch 6/10
[train] Epoch 6 mean loss = 0.0350 | mean accuracy = 0.9892
[valid] Epoch 6 mean loss = 0.2695 | mean accuracy = 0.9304

Epoch 7/10
[train] Epoch 7 mean loss = 0.0156 | mean accuracy = 0.9954
[valid] Epoch 7 mean loss = 0.2839 | mean accuracy = 0.9322

Epoch 8/10
[train] Epoch 8 mean loss = 0.0087 | mean accuracy = 0.9973
[val

# Search for hyperparameters

In [ ]:
!pip install optuna

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 364.4/364.4 kB 19.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.5/233.5 kB 22.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 8.3 MB/s eta 0:00:00


In [ ]:
#Optuna
import optuna
import time
from copy import deepcopy
import torch

# Subset the dataset for quicker testing
subset_size = 100  # Adjust this to your desired small dataset size
small_trainset = deepcopy(trainset)
small_trainset.Xtokens = trainset.Xtokens[:subset_size]
small_trainset.Ytokens = trainset.Ytokens[:subset_size]

small_validset = deepcopy(validset)
small_validset.Xtokens = validset.Xtokens[:subset_size]
small_validset.Ytokens = validset.Ytokens[:subset_size]

def objective(trial):
    # Suggest hyperparameters using Optuna
    embedding_size = trial.suggest_categorical("embedding_size", [32, 64, 128])
    hidden_size = trial.suggest_categorical("hidden_size", [64, 128, 256])
    learning_rate = trial.suggest_loguniform("learning_rate", 1e-4, 1e-2)
    batch_size = trial.suggest_categorical("batch_size", [16, 32, 64])

    print(f"\nTesting configuration: embedding_size={embedding_size}, hidden_size={hidden_size}, "
          f"learning_rate={learning_rate}, batch_size={batch_size}")

    # Create a new model
    tagger = NERtagger(small_trainset, embedding_size, hidden_size, device='cuda')

    # Train the model with the suggested hyperparameters
    tagger.train_model(
        traingenerator=small_trainset,
        validgenerator=small_validset,
        epochs=10,  # Reduced for quicker testing
        batch_size=batch_size,
        device='cuda',
        learning_rate=learning_rate
    )

    # Validate the model and retrieve metrics
    valid_loss, valid_accuracy = tagger.validate(small_validset, batch_size, device='cuda')

    # Return validation loss as the objective to minimize
    return valid_loss

# Measure time for the optimization process
start_time = time.time()

# Create a study and optimize
study = optuna.create_study(direction="minimize")  # Minimize the validation loss
study.optimize(objective, n_trials=20)  # Set number of trials

end_time = time.time()
elapsed_time = end_time - start_time

# Extrapolate based on the size of the dataset
train_size_ratio = len(trainset.Xtokens) / len(small_trainset.Xtokens)
total_time_estimate = elapsed_time * train_size_ratio

# Best result
best_result = study.best_params
best_loss = study.best_value
print("\nBest configuration:")
print(best_result)
print(f"Best validation loss: {best_loss}")
print(f"\nTime taken for small dataset: {elapsed_time:.2f} seconds")
print(f"Estimated time for full dataset: {total_time_estimate / 3600:.2f} hours")


[I 2024-11-20 19:20:40,809] A new study created in memory with name: no-name-68447067-5b2c-4bb3-9555-004b298f84f1
<ipython-input-16-6c99836de519>:21: FutureWarning: suggest_loguniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float(..., log=True) instead.
  learning_rate = trial.suggest_loguniform("learning_rate", 1e-4, 1e-2)



Testing configuration: embedding_size=32, hidden_size=256, learning_rate=0.0005212933350473318, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 2.2901 | mean accuracy = 0.1721
[valid] Epoch 1 mean loss = 2.2626 | mean accuracy = 0.3625

Epoch 2/10
[train] Epoch 2 mean loss = 2.2114 | mean accuracy = 0.7034
[valid] Epoch 2 mean loss = 2.1904 | mean accuracy = 0.7472

Epoch 3/10
[train] Epoch 3 mean loss = 2.0944 | mean accuracy = 0.8294
[valid] Epoch 3 mean loss = 2.0929 | mean accuracy = 0.7459

Epoch 4/10
[train] Epoch 4 mean loss = 1.8091 | mean accuracy = 0.8189
[valid] Epoch 4 mean loss = 1.8007 | mean accuracy = 0.7600

Epoch 5/10
[train] Epoch 5 mean loss = 1.2415 | mean accuracy = 0.8054
[valid] Epoch 5 mean loss = 1.2991 | mean accuracy = 0.7707

Epoch 6/10
[train] Epoch 6 mean loss = 0.7981 | mean accuracy = 0.8047
[valid] Epoch 6 mean loss = 1.1193 | mean accuracy = 0.7656

Epoch 7/10
[train] Epoch 7 mean loss = 0.7430 | mean accuracy = 0.8056
[valid] Epoch 7 mean loss

[I 2024-11-20 19:20:41,966] Trial 0 finished with value: 0.8417664170265198 and parameters: {'embedding_size': 32, 'hidden_size': 256, 'learning_rate': 0.0005212933350473318, 'batch_size': 64}. Best is trial 0 with value: 0.8417664170265198.


[train] Epoch 9 mean loss = 0.6438 | mean accuracy = 0.8060
[valid] Epoch 9 mean loss = 1.0591 | mean accuracy = 0.7506

Epoch 10/10
[train] Epoch 10 mean loss = 0.6071 | mean accuracy = 0.8180
[valid] Epoch 10 mean loss = 1.0580 | mean accuracy = 0.7504

Testing configuration: embedding_size=64, hidden_size=64, learning_rate=0.0023836895341926394, batch_size=16

Epoch 1/10
[train] Epoch 1 mean loss = 2.2154 | mean accuracy = 0.3679
[valid] Epoch 1 mean loss = 2.0755 | mean accuracy = 0.6899

Epoch 2/10
[train] Epoch 2 mean loss = 1.1800 | mean accuracy = 0.8399
[valid] Epoch 2 mean loss = 1.0087 | mean accuracy = 0.7852

Epoch 3/10
[train] Epoch 3 mean loss = 0.6489 | mean accuracy = 0.8212
[valid] Epoch 3 mean loss = 0.9118 | mean accuracy = 0.7946

Epoch 4/10
[train] Epoch 4 mean loss = 0.5128 | mean accuracy = 0.8599
[valid] Epoch 4 mean loss = 0.8890 | mean accuracy = 0.7955

Epoch 5/10
[train] Epoch 5 mean loss = 0.3894 | mean accuracy = 0.8888
[valid] Epoch 5 mean loss = 0.8900 

[I 2024-11-20 19:20:42,940] Trial 1 finished with value: 1.1463214755058289 and parameters: {'embedding_size': 64, 'hidden_size': 64, 'learning_rate': 0.0023836895341926394, 'batch_size': 16}. Best is trial 0 with value: 0.8417664170265198.


[train] Epoch 10 mean loss = 0.0802 | mean accuracy = 0.9870
[valid] Epoch 10 mean loss = 1.1894 | mean accuracy = 0.7955

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.00014219738640898426, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 2.3000 | mean accuracy = 0.0692
[valid] Epoch 1 mean loss = 2.2809 | mean accuracy = 0.0992

Epoch 2/10
[train] Epoch 2 mean loss = 2.2853 | mean accuracy = 0.0890
[valid] Epoch 2 mean loss = 2.2696 | mean accuracy = 0.1342

Epoch 3/10
[train] Epoch 3 mean loss = 2.2675 | mean accuracy = 0.1433
[valid] Epoch 3 mean loss = 2.2563 | mean accuracy = 0.2075

Epoch 4/10
[train] Epoch 4 mean loss = 2.2400 | mean accuracy = 0.2859
[valid] Epoch 4 mean loss = 2.2365 | mean accuracy = 0.2988

Epoch 5/10
[train] Epoch 5 mean loss = 2.1785 | mean accuracy = 0.6501
[valid] Epoch 5 mean loss = 2.1657 | mean accuracy = 0.5676

Epoch 6/10
[train] Epoch 6 mean loss = 2.0630 | mean accuracy = 0.8129
[valid] Epoch 6 mean loss = 2.0714

[I 2024-11-20 19:20:43,791] Trial 2 finished with value: 0.986338883638382 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.00014219738640898426, 'batch_size': 64}. Best is trial 0 with value: 0.8417664170265198.


[train] Epoch 9 mean loss = 1.0426 | mean accuracy = 0.8059
[valid] Epoch 9 mean loss = 1.3236 | mean accuracy = 0.7812

Epoch 10/10
[train] Epoch 10 mean loss = 0.9259 | mean accuracy = 0.8058
[valid] Epoch 10 mean loss = 1.2726 | mean accuracy = 0.7801

Testing configuration: embedding_size=32, hidden_size=128, learning_rate=0.003173919738405495, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.1571 | mean accuracy = 0.5635
[valid] Epoch 1 mean loss = 1.9738 | mean accuracy = 0.7744

Epoch 2/10
[train] Epoch 2 mean loss = 0.9212 | mean accuracy = 0.8086
[valid] Epoch 2 mean loss = 1.0367 | mean accuracy = 0.7706

Epoch 3/10
[train] Epoch 3 mean loss = 0.6497 | mean accuracy = 0.8347
[valid] Epoch 3 mean loss = 0.9746 | mean accuracy = 0.7347

Epoch 4/10
[train] Epoch 4 mean loss = 0.5611 | mean accuracy = 0.8737
[valid] Epoch 4 mean loss = 0.9188 | mean accuracy = 0.7764

Epoch 5/10
[train] Epoch 5 mean loss = 0.4390 | mean accuracy = 0.8834
[valid] Epoch 5 mean loss = 0.8968 

[I 2024-11-20 19:20:44,754] Trial 3 finished with value: 1.0789838284254074 and parameters: {'embedding_size': 32, 'hidden_size': 128, 'learning_rate': 0.003173919738405495, 'batch_size': 32}. Best is trial 0 with value: 0.8417664170265198.


[valid] Epoch 8 mean loss = 0.9944 | mean accuracy = 0.7686

Epoch 9/10
[train] Epoch 9 mean loss = 0.1352 | mean accuracy = 0.9647
[valid] Epoch 9 mean loss = 1.0974 | mean accuracy = 0.7573

Epoch 10/10
[train] Epoch 10 mean loss = 0.0978 | mean accuracy = 0.9773
[valid] Epoch 10 mean loss = 1.1318 | mean accuracy = 0.7463

Testing configuration: embedding_size=128, hidden_size=256, learning_rate=0.00021056118373457603, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.3016 | mean accuracy = 0.0975
[valid] Epoch 1 mean loss = 2.2818 | mean accuracy = 0.1472

Epoch 2/10
[train] Epoch 2 mean loss = 2.1467 | mean accuracy = 0.6098
[valid] Epoch 2 mean loss = 2.0805 | mean accuracy = 0.7128

Epoch 3/10
[train] Epoch 3 mean loss = 1.5961 | mean accuracy = 0.8218
[valid] Epoch 3 mean loss = 1.4168 | mean accuracy = 0.7756

Epoch 4/10
[train] Epoch 4 mean loss = 0.9940 | mean accuracy = 0.8177
[valid] Epoch 4 mean loss = 1.2566 | mean accuracy = 0.7753

Epoch 5/10
[train] Epoch 5 mean

[I 2024-11-20 19:20:46,423] Trial 4 finished with value: 0.8287050575017929 and parameters: {'embedding_size': 128, 'hidden_size': 256, 'learning_rate': 0.00021056118373457603, 'batch_size': 32}. Best is trial 4 with value: 0.8287050575017929.


[valid] Epoch 9 mean loss = 1.0543 | mean accuracy = 0.7853

Epoch 10/10
[train] Epoch 10 mean loss = 0.5100 | mean accuracy = 0.8757
[valid] Epoch 10 mean loss = 1.0435 | mean accuracy = 0.7805

Testing configuration: embedding_size=32, hidden_size=64, learning_rate=0.0036115471402767786, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.1895 | mean accuracy = 0.4729
[valid] Epoch 1 mean loss = 2.0572 | mean accuracy = 0.7108

Epoch 2/10
[train] Epoch 2 mean loss = 1.1091 | mean accuracy = 0.8243
[valid] Epoch 2 mean loss = 0.9719 | mean accuracy = 0.8057

Epoch 3/10
[train] Epoch 3 mean loss = 0.7096 | mean accuracy = 0.8262
[valid] Epoch 3 mean loss = 0.9287 | mean accuracy = 0.8050

Epoch 4/10
[train] Epoch 4 mean loss = 0.5709 | mean accuracy = 0.8313
[valid] Epoch 4 mean loss = 0.8835 | mean accuracy = 0.8048

Epoch 5/10
[train] Epoch 5 mean loss = 0.4749 | mean accuracy = 0.8739
[valid] Epoch 5 mean loss = 0.8840 | mean accuracy = 0.7861

Epoch 6/10
[train] Epoch 6 mean lo

[I 2024-11-20 19:20:47,324] Trial 5 finished with value: 1.0371297001838684 and parameters: {'embedding_size': 32, 'hidden_size': 64, 'learning_rate': 0.0036115471402767786, 'batch_size': 32}. Best is trial 4 with value: 0.8287050575017929.


[train] Epoch 9 mean loss = 0.1399 | mean accuracy = 0.9652
[valid] Epoch 9 mean loss = 1.0101 | mean accuracy = 0.7455

Epoch 10/10
[train] Epoch 10 mean loss = 0.1206 | mean accuracy = 0.9690
[valid] Epoch 10 mean loss = 1.0856 | mean accuracy = 0.7618

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0006365672636902998, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.2711 | mean accuracy = 0.1966
[valid] Epoch 1 mean loss = 2.2431 | mean accuracy = 0.3202

Epoch 2/10
[train] Epoch 2 mean loss = 1.9125 | mean accuracy = 0.7684
[valid] Epoch 2 mean loss = 1.6126 | mean accuracy = 0.7849

Epoch 3/10
[train] Epoch 3 mean loss = 1.0446 | mean accuracy = 0.8050
[valid] Epoch 3 mean loss = 1.1678 | mean accuracy = 0.7853

Epoch 4/10
[train] Epoch 4 mean loss = 0.7246 | mean accuracy = 0.8054
[valid] Epoch 4 mean loss = 1.0574 | mean accuracy = 0.8051

Epoch 5/10
[train] Epoch 5 mean loss = 0.6662 | mean accuracy = 0.8112
[valid] Epoch 5 mean loss = 1.0184

[I 2024-11-20 19:20:48,437] Trial 6 finished with value: 0.8238050639629364 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0006365672636902998, 'batch_size': 32}. Best is trial 6 with value: 0.8238050639629364.


[train] Epoch 10 mean loss = 0.4210 | mean accuracy = 0.8876
[valid] Epoch 10 mean loss = 0.9789 | mean accuracy = 0.7764

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.00010395347128941034, batch_size=16

Epoch 1/10
[train] Epoch 1 mean loss = 2.2966 | mean accuracy = 0.1018
[valid] Epoch 1 mean loss = 2.2903 | mean accuracy = 0.1264

Epoch 2/10
[train] Epoch 2 mean loss = 2.2497 | mean accuracy = 0.2904
[valid] Epoch 2 mean loss = 2.2357 | mean accuracy = 0.4494

Epoch 3/10
[train] Epoch 3 mean loss = 2.1754 | mean accuracy = 0.6649
[valid] Epoch 3 mean loss = 2.1752 | mean accuracy = 0.6451

Epoch 4/10
[train] Epoch 4 mean loss = 2.0882 | mean accuracy = 0.7880
[valid] Epoch 4 mean loss = 2.0979 | mean accuracy = 0.7255

Epoch 5/10
[train] Epoch 5 mean loss = 1.9646 | mean accuracy = 0.8209
[valid] Epoch 5 mean loss = 1.9835 | mean accuracy = 0.7438

Epoch 6/10
[train] Epoch 6 mean loss = 1.7376 | mean accuracy = 0.8159
[valid] Epoch 6 mean loss = 1.7565

[I 2024-11-20 19:20:49,552] Trial 7 finished with value: 1.137094599860055 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.00010395347128941034, 'batch_size': 16}. Best is trial 6 with value: 0.8238050639629364.


[train] Epoch 10 mean loss = 0.9151 | mean accuracy = 0.8213
[valid] Epoch 10 mean loss = 1.2201 | mean accuracy = 0.7751

Testing configuration: embedding_size=32, hidden_size=128, learning_rate=0.00036580791472353726, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 2.3032 | mean accuracy = 0.0516
[valid] Epoch 1 mean loss = 2.2965 | mean accuracy = 0.0444

Epoch 2/10
[train] Epoch 2 mean loss = 2.2694 | mean accuracy = 0.1870
[valid] Epoch 2 mean loss = 2.2634 | mean accuracy = 0.1919

Epoch 3/10
[train] Epoch 3 mean loss = 2.2217 | mean accuracy = 0.5086
[valid] Epoch 3 mean loss = 2.2255 | mean accuracy = 0.4965

Epoch 4/10
[train] Epoch 4 mean loss = 2.1493 | mean accuracy = 0.7479
[valid] Epoch 4 mean loss = 2.1570 | mean accuracy = 0.7273

Epoch 5/10
[train] Epoch 5 mean loss = 1.7928 | mean accuracy = 0.8128
[valid] Epoch 5 mean loss = 1.5637 | mean accuracy = 0.7751

Epoch 6/10
[train] Epoch 6 mean loss = 1.1109 | mean accuracy = 0.8050
[valid] Epoch 6 mean loss = 1.2441

[I 2024-11-20 19:20:50,335] Trial 8 finished with value: 0.8193902969360352 and parameters: {'embedding_size': 32, 'hidden_size': 128, 'learning_rate': 0.00036580791472353726, 'batch_size': 64}. Best is trial 8 with value: 0.8193902969360352.


[train] Epoch 9 mean loss = 0.7505 | mean accuracy = 0.8163
[valid] Epoch 9 mean loss = 1.0375 | mean accuracy = 0.7799

Epoch 10/10
[train] Epoch 10 mean loss = 0.7123 | mean accuracy = 0.8159
[valid] Epoch 10 mean loss = 1.0066 | mean accuracy = 0.7814

Testing configuration: embedding_size=32, hidden_size=128, learning_rate=0.00020018717938203651, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 2.2944 | mean accuracy = 0.0286
[valid] Epoch 1 mean loss = 2.2841 | mean accuracy = 0.0608

Epoch 2/10
[train] Epoch 2 mean loss = 2.2759 | mean accuracy = 0.0966
[valid] Epoch 2 mean loss = 2.2689 | mean accuracy = 0.1213

Epoch 3/10
[train] Epoch 3 mean loss = 2.2520 | mean accuracy = 0.2586
[valid] Epoch 3 mean loss = 2.2490 | mean accuracy = 0.2973

Epoch 4/10
[train] Epoch 4 mean loss = 2.2164 | mean accuracy = 0.4822
[valid] Epoch 4 mean loss = 2.2164 | mean accuracy = 0.5139

Epoch 5/10
[train] Epoch 5 mean loss = 2.1138 | mean accuracy = 0.7248
[valid] Epoch 5 mean loss = 2.071

[I 2024-11-20 19:20:51,138] Trial 9 finished with value: 0.9067936837673187 and parameters: {'embedding_size': 32, 'hidden_size': 128, 'learning_rate': 0.00020018717938203651, 'batch_size': 64}. Best is trial 8 with value: 0.8193902969360352.


[train] Epoch 10 mean loss = 0.8540 | mean accuracy = 0.8055
[valid] Epoch 10 mean loss = 1.1455 | mean accuracy = 0.7857

Testing configuration: embedding_size=128, hidden_size=64, learning_rate=0.0014580596695547177, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 2.3175 | mean accuracy = 0.0930
[valid] Epoch 1 mean loss = 2.2644 | mean accuracy = 0.1639

Epoch 2/10
[train] Epoch 2 mean loss = 2.1946 | mean accuracy = 0.3805
[valid] Epoch 2 mean loss = 2.1642 | mean accuracy = 0.4420

Epoch 3/10
[train] Epoch 3 mean loss = 2.0176 | mean accuracy = 0.6816
[valid] Epoch 3 mean loss = 2.0101 | mean accuracy = 0.6495

Epoch 4/10
[train] Epoch 4 mean loss = 1.6550 | mean accuracy = 0.8278
[valid] Epoch 4 mean loss = 1.6751 | mean accuracy = 0.7487

Epoch 5/10
[train] Epoch 5 mean loss = 0.9389 | mean accuracy = 0.8460
[valid] Epoch 5 mean loss = 1.0950 | mean accuracy = 0.7754

Epoch 6/10
[train] Epoch 6 mean loss = 0.5834 | mean accuracy = 0.8426
[valid] Epoch 6 mean loss = 1.0296 

[I 2024-11-20 19:20:52,588] Trial 10 finished with value: 0.8495599031448364 and parameters: {'embedding_size': 128, 'hidden_size': 64, 'learning_rate': 0.0014580596695547177, 'batch_size': 64}. Best is trial 8 with value: 0.8193902969360352.


[valid] Epoch 9 mean loss = 0.9971 | mean accuracy = 0.7610

Epoch 10/10
[train] Epoch 10 mean loss = 0.1770 | mean accuracy = 0.9562
[valid] Epoch 10 mean loss = 0.9987 | mean accuracy = 0.7673

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.000609549418897796, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.2760 | mean accuracy = 0.1560
[valid] Epoch 1 mean loss = 2.2437 | mean accuracy = 0.3351

Epoch 2/10
[train] Epoch 2 mean loss = 1.9923 | mean accuracy = 0.7515
[valid] Epoch 2 mean loss = 1.7693 | mean accuracy = 0.7776

Epoch 3/10
[train] Epoch 3 mean loss = 1.1112 | mean accuracy = 0.8208
[valid] Epoch 3 mean loss = 1.1983 | mean accuracy = 0.7852

Epoch 4/10
[train] Epoch 4 mean loss = 0.7477 | mean accuracy = 0.8200
[valid] Epoch 4 mean loss = 1.0803 | mean accuracy = 0.7845

Epoch 5/10
[train] Epoch 5 mean loss = 0.6798 | mean accuracy = 0.8362
[valid] Epoch 5 mean loss = 1.0324 | mean accuracy = 0.7997

Epoch 6/10
[train] Epoch 6 mean lo

[I 2024-11-20 19:20:53,918] Trial 11 finished with value: 0.8099141269922256 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.000609549418897796, 'batch_size': 32}. Best is trial 11 with value: 0.8099141269922256.


[train] Epoch 10 mean loss = 0.4081 | mean accuracy = 0.8852
[valid] Epoch 10 mean loss = 0.9767 | mean accuracy = 0.7779

Testing configuration: embedding_size=32, hidden_size=128, learning_rate=0.00038012851497897356, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 2.2915 | mean accuracy = 0.0848
[valid] Epoch 1 mean loss = 2.2747 | mean accuracy = 0.2335

Epoch 2/10
[train] Epoch 2 mean loss = 2.2570 | mean accuracy = 0.3055
[valid] Epoch 2 mean loss = 2.2413 | mean accuracy = 0.4836

Epoch 3/10
[train] Epoch 3 mean loss = 2.2104 | mean accuracy = 0.6063
[valid] Epoch 3 mean loss = 2.2020 | mean accuracy = 0.6447

Epoch 4/10
[train] Epoch 4 mean loss = 2.1361 | mean accuracy = 0.7667
[valid] Epoch 4 mean loss = 2.1301 | mean accuracy = 0.7222

Epoch 5/10
[train] Epoch 5 mean loss = 1.7599 | mean accuracy = 0.8037
[valid] Epoch 5 mean loss = 1.4856 | mean accuracy = 0.7703

Epoch 6/10
[train] Epoch 6 mean loss = 1.1122 | mean accuracy = 0.8048
[valid] Epoch 6 mean loss = 1.2575

[I 2024-11-20 19:20:54,704] Trial 12 finished with value: 0.8332357704639435 and parameters: {'embedding_size': 32, 'hidden_size': 128, 'learning_rate': 0.00038012851497897356, 'batch_size': 64}. Best is trial 11 with value: 0.8099141269922256.


[train] Epoch 9 mean loss = 0.7577 | mean accuracy = 0.8063
[valid] Epoch 9 mean loss = 1.0701 | mean accuracy = 0.7904

Epoch 10/10
[train] Epoch 10 mean loss = 0.7251 | mean accuracy = 0.8049
[valid] Epoch 10 mean loss = 1.0484 | mean accuracy = 0.7903

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0010851759746442, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.2723 | mean accuracy = 0.2115
[valid] Epoch 1 mean loss = 2.2261 | mean accuracy = 0.3575

Epoch 2/10
[train] Epoch 2 mean loss = 1.5116 | mean accuracy = 0.8151
[valid] Epoch 2 mean loss = 1.1835 | mean accuracy = 0.7682

Epoch 3/10
[train] Epoch 3 mean loss = 0.7192 | mean accuracy = 0.8107
[valid] Epoch 3 mean loss = 1.0349 | mean accuracy = 0.7784

Epoch 4/10
[train] Epoch 4 mean loss = 0.6365 | mean accuracy = 0.8369
[valid] Epoch 4 mean loss = 0.9773 | mean accuracy = 0.7782

Epoch 5/10
[train] Epoch 5 mean loss = 0.5321 | mean accuracy = 0.8582
[valid] Epoch 5 mean loss = 0.9577 | 

[I 2024-11-20 19:20:55,812] Trial 13 finished with value: 0.8629544824361801 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0010851759746442, 'batch_size': 32}. Best is trial 11 with value: 0.8099141269922256.


[train] Epoch 9 mean loss = 0.2392 | mean accuracy = 0.9312
[valid] Epoch 9 mean loss = 0.9814 | mean accuracy = 0.7152

Epoch 10/10
[train] Epoch 10 mean loss = 0.1885 | mean accuracy = 0.9550
[valid] Epoch 10 mean loss = 1.0029 | mean accuracy = 0.7265

Testing configuration: embedding_size=128, hidden_size=256, learning_rate=0.006769400276704751, batch_size=16

Epoch 1/10
[train] Epoch 1 mean loss = 1.2833 | mean accuracy = 0.7404
[valid] Epoch 1 mean loss = 1.0422 | mean accuracy = 0.7899

Epoch 2/10
[train] Epoch 2 mean loss = 0.4853 | mean accuracy = 0.8608
[valid] Epoch 2 mean loss = 0.8991 | mean accuracy = 0.7611

Epoch 3/10
[train] Epoch 3 mean loss = 0.1296 | mean accuracy = 0.9705
[valid] Epoch 3 mean loss = 0.9740 | mean accuracy = 0.7695

Epoch 4/10
[train] Epoch 4 mean loss = 0.0339 | mean accuracy = 0.9896
[valid] Epoch 4 mean loss = 1.0572 | mean accuracy = 0.7700

Epoch 5/10
[train] Epoch 5 mean loss = 0.0209 | mean accuracy = 0.9940
[valid] Epoch 5 mean loss = 1.2991

[I 2024-11-20 19:20:57,218] Trial 14 finished with value: 1.3474999581064497 and parameters: {'embedding_size': 128, 'hidden_size': 256, 'learning_rate': 0.006769400276704751, 'batch_size': 16}. Best is trial 11 with value: 0.8099141269922256.


[train] Epoch 10 mean loss = 0.0004 | mean accuracy = 1.0000
[valid] Epoch 10 mean loss = 1.3445 | mean accuracy = 0.7630

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0003831070427482731, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.2215 | mean accuracy = 0.5021
[valid] Epoch 1 mean loss = 2.1876 | mean accuracy = 0.6442

Epoch 2/10
[train] Epoch 2 mean loss = 2.0523 | mean accuracy = 0.7895
[valid] Epoch 2 mean loss = 1.9548 | mean accuracy = 0.7796

Epoch 3/10
[train] Epoch 3 mean loss = 1.3635 | mean accuracy = 0.8366
[valid] Epoch 3 mean loss = 1.2861 | mean accuracy = 0.8054

Epoch 4/10
[train] Epoch 4 mean loss = 0.9162 | mean accuracy = 0.8212
[valid] Epoch 4 mean loss = 1.1665 | mean accuracy = 0.8055

Epoch 5/10
[train] Epoch 5 mean loss = 0.7720 | mean accuracy = 0.8202
[valid] Epoch 5 mean loss = 1.1064 | mean accuracy = 0.8057

Epoch 6/10
[train] Epoch 6 mean loss = 0.7101 | mean accuracy = 0.8269
[valid] Epoch 6 mean loss = 1.0731 

[I 2024-11-20 19:20:58,304] Trial 15 finished with value: 0.7881987541913986 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0003831070427482731, 'batch_size': 32}. Best is trial 15 with value: 0.7881987541913986.


[valid] Epoch 10 mean loss = 1.0062 | mean accuracy = 0.8011

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0007323806892369036, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.3013 | mean accuracy = 0.0757
[valid] Epoch 1 mean loss = 2.2352 | mean accuracy = 0.3141

Epoch 2/10
[train] Epoch 2 mean loss = 1.8012 | mean accuracy = 0.7427
[valid] Epoch 2 mean loss = 1.2823 | mean accuracy = 0.7954

Epoch 3/10
[train] Epoch 3 mean loss = 0.9711 | mean accuracy = 0.8161
[valid] Epoch 3 mean loss = 1.1075 | mean accuracy = 0.8048

Epoch 4/10
[train] Epoch 4 mean loss = 0.7199 | mean accuracy = 0.8167
[valid] Epoch 4 mean loss = 1.0057 | mean accuracy = 0.8051

Epoch 5/10
[train] Epoch 5 mean loss = 0.6498 | mean accuracy = 0.8368
[valid] Epoch 5 mean loss = 0.9681 | mean accuracy = 0.8059

Epoch 6/10
[train] Epoch 6 mean loss = 0.5945 | mean accuracy = 0.8411
[valid] Epoch 6 mean loss = 0.9423 | mean accuracy = 0.8046

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 19:20:59,385] Trial 16 finished with value: 0.7878151386976242 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0007323806892369036, 'batch_size': 32}. Best is trial 16 with value: 0.7878151386976242.


[train] Epoch 9 mean loss = 0.4378 | mean accuracy = 0.8797
[valid] Epoch 9 mean loss = 0.9166 | mean accuracy = 0.7886

Epoch 10/10
[train] Epoch 10 mean loss = 0.3814 | mean accuracy = 0.8936
[valid] Epoch 10 mean loss = 0.9282 | mean accuracy = 0.7866

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0016596900409684482, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.2456 | mean accuracy = 0.3190
[valid] Epoch 1 mean loss = 2.1458 | mean accuracy = 0.6637

Epoch 2/10
[train] Epoch 2 mean loss = 1.2653 | mean accuracy = 0.8171
[valid] Epoch 2 mean loss = 1.0582 | mean accuracy = 0.7609

Epoch 3/10
[train] Epoch 3 mean loss = 0.6937 | mean accuracy = 0.8209
[valid] Epoch 3 mean loss = 0.9618 | mean accuracy = 0.7659

Epoch 4/10
[train] Epoch 4 mean loss = 0.5586 | mean accuracy = 0.8458
[valid] Epoch 4 mean loss = 0.9206 | mean accuracy = 0.7994

Epoch 5/10
[train] Epoch 5 mean loss = 0.4532 | mean accuracy = 0.8781
[valid] Epoch 5 mean loss = 0.8885

[I 2024-11-20 19:21:00,452] Trial 17 finished with value: 1.0826634019613266 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0016596900409684482, 'batch_size': 32}. Best is trial 16 with value: 0.7878151386976242.


[valid] Epoch 10 mean loss = 1.1155 | mean accuracy = 0.7907

Testing configuration: embedding_size=64, hidden_size=64, learning_rate=0.0008725106985317709, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.2408 | mean accuracy = 0.2203
[valid] Epoch 1 mean loss = 2.2176 | mean accuracy = 0.3454

Epoch 2/10
[train] Epoch 2 mean loss = 1.9905 | mean accuracy = 0.7824
[valid] Epoch 2 mean loss = 1.8340 | mean accuracy = 0.7800

Epoch 3/10
[train] Epoch 3 mean loss = 1.1241 | mean accuracy = 0.8156
[valid] Epoch 3 mean loss = 1.1543 | mean accuracy = 0.7798

Epoch 4/10
[train] Epoch 4 mean loss = 0.7228 | mean accuracy = 0.8157
[valid] Epoch 4 mean loss = 1.0644 | mean accuracy = 0.8007

Epoch 5/10
[train] Epoch 5 mean loss = 0.6386 | mean accuracy = 0.8304
[valid] Epoch 5 mean loss = 1.0264 | mean accuracy = 0.8003

Epoch 6/10
[train] Epoch 6 mean loss = 0.5774 | mean accuracy = 0.8312
[valid] Epoch 6 mean loss = 1.0021 | mean accuracy = 0.7971

Epoch 7/10
[train] Epoch 7 mean loss

[I 2024-11-20 19:21:01,484] Trial 18 finished with value: 0.8394912332296371 and parameters: {'embedding_size': 64, 'hidden_size': 64, 'learning_rate': 0.0008725106985317709, 'batch_size': 32}. Best is trial 16 with value: 0.7878151386976242.


[train] Epoch 9 mean loss = 0.4058 | mean accuracy = 0.8797
[valid] Epoch 9 mean loss = 0.9881 | mean accuracy = 0.7700

Epoch 10/10
[train] Epoch 10 mean loss = 0.3469 | mean accuracy = 0.8934
[valid] Epoch 10 mean loss = 1.0057 | mean accuracy = 0.7635

Testing configuration: embedding_size=64, hidden_size=256, learning_rate=0.0002878359641906546, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 2.2489 | mean accuracy = 0.4788
[valid] Epoch 1 mean loss = 2.2234 | mean accuracy = 0.6789

Epoch 2/10
[train] Epoch 2 mean loss = 2.0642 | mean accuracy = 0.7972
[valid] Epoch 2 mean loss = 1.9284 | mean accuracy = 0.7855

Epoch 3/10
[train] Epoch 3 mean loss = 1.2904 | mean accuracy = 0.8248
[valid] Epoch 3 mean loss = 1.2705 | mean accuracy = 0.7847

Epoch 4/10
[train] Epoch 4 mean loss = 0.8934 | mean accuracy = 0.8257
[valid] Epoch 4 mean loss = 1.1347 | mean accuracy = 0.8046

Epoch 5/10
[train] Epoch 5 mean loss = 0.7790 | mean accuracy = 0.8364
[valid] Epoch 5 mean loss = 1.0763

[I 2024-11-20 19:21:02,922] Trial 19 finished with value: 0.7641179263591766 and parameters: {'embedding_size': 64, 'hidden_size': 256, 'learning_rate': 0.0002878359641906546, 'batch_size': 32}. Best is trial 19 with value: 0.7641179263591766.


[valid] Epoch 10 mean loss = 0.9740 | mean accuracy = 0.7951

Best configuration:
{'embedding_size': 64, 'hidden_size': 256, 'learning_rate': 0.0002878359641906546, 'batch_size': 32}
Best validation loss: 0.7641179263591766

Time taken for small dataset: 22.12 seconds
Estimated time for full dataset: 0.86 hours


In [ ]:
import optuna
import time
import torch

# Load the full dataset
trainset = DataGenerator('eng.train')  # Full training dataset
validset = DataGenerator('eng.testa', parentgenerator=trainset)  # Validation dataset

def objective(trial):
    # Suggest hyperparameters using Optuna
    embedding_size = trial.suggest_categorical("embedding_size", [32, 64, 128])
    hidden_size = trial.suggest_categorical("hidden_size", [64, 128, 256])
    learning_rate = trial.suggest_loguniform("learning_rate", 1e-4, 1e-2)
    batch_size = trial.suggest_categorical("batch_size", [16, 32, 64])

    print(f"\nTesting configuration: embedding_size={embedding_size}, hidden_size={hidden_size}, "
          f"learning_rate={learning_rate}, batch_size={batch_size}")

    # Create a new model
    tagger = NERtagger(trainset, embedding_size, hidden_size, device='cuda')

    # Train the model with the suggested hyperparameters
    tagger.train_model(
        traingenerator=trainset,
        validgenerator=validset,
        epochs=10,  # Number of epochs
        batch_size=batch_size,
        device='cuda',
        learning_rate=learning_rate
    )

    # Validate the model and retrieve metrics
    valid_loss, valid_accuracy = tagger.validate(validset, batch_size, device='cuda')

    # Return validation loss as the objective to minimize
    return valid_loss

# Measure time for the optimization process
start_time = time.time()

# Create a study and optimize
study = optuna.create_study(direction="minimize")  # Minimize the validation loss
study.optimize(objective, n_trials=20)  # Number of trials

end_time = time.time()
elapsed_time = end_time - start_time

# Best result
best_result = study.best_params
best_loss = study.best_value
print("\nBest configuration:")
print(best_result)
print(f"Best validation loss: {best_loss}")
print(f"\nTotal time taken: {elapsed_time:.2f} seconds")

[I 2024-11-20 19:21:16,237] A new study created in memory with name: no-name-74c36813-6f9f-4337-8b3f-d622490fb0d4
<ipython-input-17-c298c45a410c>:13: FutureWarning: suggest_loguniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float(..., log=True) instead.
  learning_rate = trial.suggest_loguniform("learning_rate", 1e-4, 1e-2)



Testing configuration: embedding_size=128, hidden_size=128, learning_rate=0.0060505333502602655, batch_size=16

Epoch 1/10
[train] Epoch 1 mean loss = 0.2839 | mean accuracy = 0.9166
[valid] Epoch 1 mean loss = 0.2285 | mean accuracy = 0.9303

Epoch 2/10
[train] Epoch 2 mean loss = 0.1050 | mean accuracy = 0.9678
[valid] Epoch 2 mean loss = 0.2113 | mean accuracy = 0.9464

Epoch 3/10
[train] Epoch 3 mean loss = 0.1044 | mean accuracy = 0.9715
[valid] Epoch 3 mean loss = 0.3355 | mean accuracy = 0.9073

Epoch 4/10
[train] Epoch 4 mean loss = 0.0753 | mean accuracy = 0.9792
[valid] Epoch 4 mean loss = 0.3110 | mean accuracy = 0.9222

Epoch 5/10
[train] Epoch 5 mean loss = 0.0579 | mean accuracy = 0.9846
[valid] Epoch 5 mean loss = 0.3496 | mean accuracy = 0.9238

Epoch 6/10
[train] Epoch 6 mean loss = 0.0464 | mean accuracy = 0.9878
[valid] Epoch 6 mean loss = 0.3535 | mean accuracy = 0.9308

Epoch 7/10
[train] Epoch 7 mean loss = 0.0422 | mean accuracy = 0.9892
[valid] Epoch 7 mean los

[I 2024-11-20 19:27:59,652] Trial 0 finished with value: 0.42292219349274446 and parameters: {'embedding_size': 128, 'hidden_size': 128, 'learning_rate': 0.0060505333502602655, 'batch_size': 16}. Best is trial 0 with value: 0.42292219349274446.


[valid] Epoch 10 mean loss = 0.4216 | mean accuracy = 0.9398

Testing configuration: embedding_size=128, hidden_size=128, learning_rate=0.00516785550973882, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 0.3524 | mean accuracy = 0.8949
[valid] Epoch 1 mean loss = 0.2529 | mean accuracy = 0.9222

Epoch 2/10
[train] Epoch 2 mean loss = 0.1044 | mean accuracy = 0.9671
[valid] Epoch 2 mean loss = 0.2062 | mean accuracy = 0.9430

Epoch 3/10
[train] Epoch 3 mean loss = 0.0225 | mean accuracy = 0.9935
[valid] Epoch 3 mean loss = 0.2409 | mean accuracy = 0.9484

Epoch 4/10
[train] Epoch 4 mean loss = 0.0259 | mean accuracy = 0.9921
[valid] Epoch 4 mean loss = 0.2815 | mean accuracy = 0.9443

Epoch 5/10
[train] Epoch 5 mean loss = 0.0749 | mean accuracy = 0.9797
[valid] Epoch 5 mean loss = 0.3596 | mean accuracy = 0.9232

Epoch 6/10
[train] Epoch 6 mean loss = 0.0530 | mean accuracy = 0.9851
[valid] Epoch 6 mean loss = 0.3535 | mean accuracy = 0.9243

Epoch 7/10
[train] Epoch 7 mean loss

[I 2024-11-20 19:33:06,467] Trial 1 finished with value: 0.5394770292674794 and parameters: {'embedding_size': 128, 'hidden_size': 128, 'learning_rate': 0.00516785550973882, 'batch_size': 32}. Best is trial 0 with value: 0.42292219349274446.


[valid] Epoch 10 mean loss = 0.5359 | mean accuracy = 0.9460

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0023323062867411665, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 0.6980 | mean accuracy = 0.8030
[valid] Epoch 1 mean loss = 0.5021 | mean accuracy = 0.8395

Epoch 2/10
[train] Epoch 2 mean loss = 0.2856 | mean accuracy = 0.9107
[valid] Epoch 2 mean loss = 0.2796 | mean accuracy = 0.9060

Epoch 3/10
[train] Epoch 3 mean loss = 0.1274 | mean accuracy = 0.9605
[valid] Epoch 3 mean loss = 0.2044 | mean accuracy = 0.9327

Epoch 4/10
[train] Epoch 4 mean loss = 0.0501 | mean accuracy = 0.9853
[valid] Epoch 4 mean loss = 0.1895 | mean accuracy = 0.9414

Epoch 5/10
[train] Epoch 5 mean loss = 0.0576 | mean accuracy = 0.9819
[valid] Epoch 5 mean loss = 0.2011 | mean accuracy = 0.9434

Epoch 6/10
[train] Epoch 6 mean loss = 0.0305 | mean accuracy = 0.9907
[valid] Epoch 6 mean loss = 0.1945 | mean accuracy = 0.9471

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 19:37:32,113] Trial 2 finished with value: 0.28188638622854273 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0023323062867411665, 'batch_size': 64}. Best is trial 2 with value: 0.28188638622854273.


[valid] Epoch 10 mean loss = 0.2789 | mean accuracy = 0.9493

Testing configuration: embedding_size=64, hidden_size=256, learning_rate=0.00012812756195067034, batch_size=16

Epoch 1/10
[train] Epoch 1 mean loss = 0.8857 | mean accuracy = 0.7730
[valid] Epoch 1 mean loss = 0.7721 | mean accuracy = 0.7981

Epoch 2/10
[train] Epoch 2 mean loss = 0.5589 | mean accuracy = 0.8367
[valid] Epoch 2 mean loss = 0.5750 | mean accuracy = 0.8261

Epoch 3/10
[train] Epoch 3 mean loss = 0.3922 | mean accuracy = 0.8829
[valid] Epoch 3 mean loss = 0.4638 | mean accuracy = 0.8528

Epoch 4/10
[train] Epoch 4 mean loss = 0.2852 | mean accuracy = 0.9119
[valid] Epoch 4 mean loss = 0.4107 | mean accuracy = 0.8631

Epoch 5/10
[train] Epoch 5 mean loss = 0.2103 | mean accuracy = 0.9341
[valid] Epoch 5 mean loss = 0.4031 | mean accuracy = 0.8642

Epoch 6/10
[train] Epoch 6 mean loss = 0.1532 | mean accuracy = 0.9526
[valid] Epoch 6 mean loss = 0.3997 | mean accuracy = 0.8704

Epoch 7/10
[train] Epoch 7 mean lo

[I 2024-11-20 19:44:41,954] Trial 3 finished with value: 0.483205793443702 and parameters: {'embedding_size': 64, 'hidden_size': 256, 'learning_rate': 0.00012812756195067034, 'batch_size': 16}. Best is trial 2 with value: 0.28188638622854273.



Testing configuration: embedding_size=64, hidden_size=64, learning_rate=0.0024162771804004833, batch_size=16

Epoch 1/10
[train] Epoch 1 mean loss = 0.4731 | mean accuracy = 0.8593
[valid] Epoch 1 mean loss = 0.3553 | mean accuracy = 0.8878

Epoch 2/10
[train] Epoch 2 mean loss = 0.1759 | mean accuracy = 0.9455
[valid] Epoch 2 mean loss = 0.2221 | mean accuracy = 0.9322

Epoch 3/10
[train] Epoch 3 mean loss = 0.1005 | mean accuracy = 0.9684
[valid] Epoch 3 mean loss = 0.2494 | mean accuracy = 0.9349

Epoch 4/10
[train] Epoch 4 mean loss = 0.0480 | mean accuracy = 0.9852
[valid] Epoch 4 mean loss = 0.2482 | mean accuracy = 0.9433

Epoch 5/10
[train] Epoch 5 mean loss = 0.0253 | mean accuracy = 0.9922
[valid] Epoch 5 mean loss = 0.2700 | mean accuracy = 0.9504

Epoch 6/10
[train] Epoch 6 mean loss = 0.0149 | mean accuracy = 0.9954
[valid] Epoch 6 mean loss = 0.2416 | mean accuracy = 0.9487

Epoch 7/10
[train] Epoch 7 mean loss = 0.0108 | mean accuracy = 0.9967
[valid] Epoch 7 mean loss 

[I 2024-11-20 19:50:25,836] Trial 4 finished with value: 0.4450937364321469 and parameters: {'embedding_size': 64, 'hidden_size': 64, 'learning_rate': 0.0024162771804004833, 'batch_size': 16}. Best is trial 2 with value: 0.28188638622854273.


[valid] Epoch 10 mean loss = 0.4415 | mean accuracy = 0.9522

Testing configuration: embedding_size=128, hidden_size=64, learning_rate=0.0014274728479534019, batch_size=16

Epoch 1/10
[train] Epoch 1 mean loss = 0.4912 | mean accuracy = 0.8579
[valid] Epoch 1 mean loss = 0.3528 | mean accuracy = 0.8959

Epoch 2/10
[train] Epoch 2 mean loss = 0.1733 | mean accuracy = 0.9468
[valid] Epoch 2 mean loss = 0.2726 | mean accuracy = 0.9070

Epoch 3/10
[train] Epoch 3 mean loss = 0.0939 | mean accuracy = 0.9712
[valid] Epoch 3 mean loss = 0.3535 | mean accuracy = 0.8947

Epoch 4/10
[train] Epoch 4 mean loss = 0.0375 | mean accuracy = 0.9881
[valid] Epoch 4 mean loss = 0.5003 | mean accuracy = 0.8959

Epoch 5/10
[train] Epoch 5 mean loss = 0.0179 | mean accuracy = 0.9947
[valid] Epoch 5 mean loss = 0.5540 | mean accuracy = 0.8999

Epoch 6/10
[train] Epoch 6 mean loss = 0.0096 | mean accuracy = 0.9972
[valid] Epoch 6 mean loss = 0.6006 | mean accuracy = 0.9023

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 19:56:54,095] Trial 5 finished with value: 1.0460843174334835 and parameters: {'embedding_size': 128, 'hidden_size': 64, 'learning_rate': 0.0014274728479534019, 'batch_size': 16}. Best is trial 2 with value: 0.28188638622854273.



Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0005015110812981718, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 0.9971 | mean accuracy = 0.7516
[valid] Epoch 1 mean loss = 0.8089 | mean accuracy = 0.7939

Epoch 2/10
[train] Epoch 2 mean loss = 0.5714 | mean accuracy = 0.8281
[valid] Epoch 2 mean loss = 0.5209 | mean accuracy = 0.8453

Epoch 3/10
[train] Epoch 3 mean loss = 0.3630 | mean accuracy = 0.8880
[valid] Epoch 3 mean loss = 0.3891 | mean accuracy = 0.8779

Epoch 4/10
[train] Epoch 4 mean loss = 0.2465 | mean accuracy = 0.9232
[valid] Epoch 4 mean loss = 0.3030 | mean accuracy = 0.9015

Epoch 5/10
[train] Epoch 5 mean loss = 0.1667 | mean accuracy = 0.9471
[valid] Epoch 5 mean loss = 0.2900 | mean accuracy = 0.9016

Epoch 6/10
[train] Epoch 6 mean loss = 0.0861 | mean accuracy = 0.9738
[valid] Epoch 6 mean loss = 0.2467 | mean accuracy = 0.9191

Epoch 7/10
[train] Epoch 7 mean loss = 0.0392 | mean accuracy = 0.9885
[valid] Epoch 7 mean loss

[I 2024-11-20 20:01:20,297] Trial 6 finished with value: 0.2900930984931834 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0005015110812981718, 'batch_size': 64}. Best is trial 2 with value: 0.28188638622854273.


[valid] Epoch 10 mean loss = 0.2733 | mean accuracy = 0.9372

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.002832076974710779, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 0.5185 | mean accuracy = 0.8487
[valid] Epoch 1 mean loss = 0.4105 | mean accuracy = 0.8600

Epoch 2/10
[train] Epoch 2 mean loss = 0.1980 | mean accuracy = 0.9383
[valid] Epoch 2 mean loss = 0.2619 | mean accuracy = 0.9094

Epoch 3/10
[train] Epoch 3 mean loss = 0.0778 | mean accuracy = 0.9765
[valid] Epoch 3 mean loss = 0.2353 | mean accuracy = 0.9201

Epoch 4/10
[train] Epoch 4 mean loss = 0.0297 | mean accuracy = 0.9914
[valid] Epoch 4 mean loss = 0.2085 | mean accuracy = 0.9374

Epoch 5/10
[train] Epoch 5 mean loss = 0.0593 | mean accuracy = 0.9816
[valid] Epoch 5 mean loss = 0.5088 | mean accuracy = 0.8831

Epoch 6/10
[train] Epoch 6 mean loss = 0.0345 | mean accuracy = 0.9896
[valid] Epoch 6 mean loss = 0.6396 | mean accuracy = 0.8888

Epoch 7/10
[train] Epoch 7 mean loss

[I 2024-11-20 20:05:47,054] Trial 7 finished with value: 0.607401345874749 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.002832076974710779, 'batch_size': 32}. Best is trial 2 with value: 0.28188638622854273.


[valid] Epoch 10 mean loss = 0.5954 | mean accuracy = 0.9107

Testing configuration: embedding_size=128, hidden_size=128, learning_rate=0.00972651280210201, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 0.3759 | mean accuracy = 0.8899
[valid] Epoch 1 mean loss = 0.2522 | mean accuracy = 0.9211

Epoch 2/10
[train] Epoch 2 mean loss = 0.1112 | mean accuracy = 0.9657
[valid] Epoch 2 mean loss = 0.2439 | mean accuracy = 0.9293

Epoch 3/10
[train] Epoch 3 mean loss = 0.0358 | mean accuracy = 0.9894
[valid] Epoch 3 mean loss = 0.2229 | mean accuracy = 0.9443

Epoch 4/10
[train] Epoch 4 mean loss = 0.0453 | mean accuracy = 0.9870
[valid] Epoch 4 mean loss = 0.2698 | mean accuracy = 0.9368

Epoch 5/10
[train] Epoch 5 mean loss = 0.1062 | mean accuracy = 0.9759
[valid] Epoch 5 mean loss = 0.4487 | mean accuracy = 0.8908

Epoch 6/10
[train] Epoch 6 mean loss = 0.1002 | mean accuracy = 0.9767
[valid] Epoch 6 mean loss = 0.3873 | mean accuracy = 0.8979

Epoch 7/10
[train] Epoch 7 mean loss

[I 2024-11-20 20:10:53,133] Trial 8 finished with value: 0.4960346952372906 and parameters: {'embedding_size': 128, 'hidden_size': 128, 'learning_rate': 0.00972651280210201, 'batch_size': 64}. Best is trial 2 with value: 0.28188638622854273.


[valid] Epoch 10 mean loss = 0.4906 | mean accuracy = 0.8952

Testing configuration: embedding_size=32, hidden_size=256, learning_rate=0.0009029264195416358, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 0.7244 | mean accuracy = 0.7933
[valid] Epoch 1 mean loss = 0.6281 | mean accuracy = 0.8090

Epoch 2/10
[train] Epoch 2 mean loss = 0.4549 | mean accuracy = 0.8605
[valid] Epoch 2 mean loss = 0.4670 | mean accuracy = 0.8439

Epoch 3/10
[train] Epoch 3 mean loss = 0.3086 | mean accuracy = 0.9035
[valid] Epoch 3 mean loss = 0.3455 | mean accuracy = 0.8868

Epoch 4/10
[train] Epoch 4 mean loss = 0.2085 | mean accuracy = 0.9346
[valid] Epoch 4 mean loss = 0.2798 | mean accuracy = 0.9062

Epoch 5/10
[train] Epoch 5 mean loss = 0.1503 | mean accuracy = 0.9521
[valid] Epoch 5 mean loss = 0.2805 | mean accuracy = 0.9054

Epoch 6/10
[train] Epoch 6 mean loss = 0.0745 | mean accuracy = 0.9767
[valid] Epoch 6 mean loss = 0.2732 | mean accuracy = 0.9177

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 20:16:10,532] Trial 9 finished with value: 0.33900219822923344 and parameters: {'embedding_size': 32, 'hidden_size': 256, 'learning_rate': 0.0009029264195416358, 'batch_size': 32}. Best is trial 2 with value: 0.28188638622854273.


[valid] Epoch 10 mean loss = 0.3370 | mean accuracy = 0.9217

Testing configuration: embedding_size=32, hidden_size=64, learning_rate=0.0003414176792983604, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 1.2129 | mean accuracy = 0.7119
[valid] Epoch 1 mean loss = 0.9519 | mean accuracy = 0.7706

Epoch 2/10
[train] Epoch 2 mean loss = 0.7796 | mean accuracy = 0.7770
[valid] Epoch 2 mean loss = 0.7475 | mean accuracy = 0.7980

Epoch 3/10
[train] Epoch 3 mean loss = 0.6021 | mean accuracy = 0.8189
[valid] Epoch 3 mean loss = 0.6002 | mean accuracy = 0.8241

Epoch 4/10
[train] Epoch 4 mean loss = 0.4584 | mean accuracy = 0.8588
[valid] Epoch 4 mean loss = 0.4886 | mean accuracy = 0.8508

Epoch 5/10
[train] Epoch 5 mean loss = 0.3281 | mean accuracy = 0.8976
[valid] Epoch 5 mean loss = 0.4221 | mean accuracy = 0.8608

Epoch 6/10
[train] Epoch 6 mean loss = 0.2357 | mean accuracy = 0.9260
[valid] Epoch 6 mean loss = 0.3696 | mean accuracy = 0.8720

Epoch 7/10
[train] Epoch 7 mean loss

[I 2024-11-20 20:20:23,964] Trial 10 finished with value: 0.3406760724736195 and parameters: {'embedding_size': 32, 'hidden_size': 64, 'learning_rate': 0.0003414176792983604, 'batch_size': 64}. Best is trial 2 with value: 0.28188638622854273.


[valid] Epoch 10 mean loss = 0.3399 | mean accuracy = 0.8905

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0005420612939490512, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 0.9773 | mean accuracy = 0.7453
[valid] Epoch 1 mean loss = 0.7815 | mean accuracy = 0.7941

Epoch 2/10
[train] Epoch 2 mean loss = 0.5558 | mean accuracy = 0.8354
[valid] Epoch 2 mean loss = 0.5106 | mean accuracy = 0.8431

Epoch 3/10
[train] Epoch 3 mean loss = 0.3547 | mean accuracy = 0.8926
[valid] Epoch 3 mean loss = 0.3815 | mean accuracy = 0.8808

Epoch 4/10
[train] Epoch 4 mean loss = 0.2387 | mean accuracy = 0.9261
[valid] Epoch 4 mean loss = 0.2968 | mean accuracy = 0.9081

Epoch 5/10
[train] Epoch 5 mean loss = 0.1592 | mean accuracy = 0.9498
[valid] Epoch 5 mean loss = 0.2432 | mean accuracy = 0.9224

Epoch 6/10
[train] Epoch 6 mean loss = 0.0789 | mean accuracy = 0.9759
[valid] Epoch 6 mean loss = 0.2081 | mean accuracy = 0.9376

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 20:24:51,437] Trial 11 finished with value: 0.26232793138307686 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0005420612939490512, 'batch_size': 64}. Best is trial 11 with value: 0.26232793138307686.


[valid] Epoch 10 mean loss = 0.2539 | mean accuracy = 0.9450

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0004520724638002848, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 1.0100 | mean accuracy = 0.7579
[valid] Epoch 1 mean loss = 0.8009 | mean accuracy = 0.7845

Epoch 2/10
[train] Epoch 2 mean loss = 0.5925 | mean accuracy = 0.8241
[valid] Epoch 2 mean loss = 0.5272 | mean accuracy = 0.8427

Epoch 3/10
[train] Epoch 3 mean loss = 0.3861 | mean accuracy = 0.8810
[valid] Epoch 3 mean loss = 0.4046 | mean accuracy = 0.8739

Epoch 4/10
[train] Epoch 4 mean loss = 0.2696 | mean accuracy = 0.9164
[valid] Epoch 4 mean loss = 0.3262 | mean accuracy = 0.8953

Epoch 5/10
[train] Epoch 5 mean loss = 0.1853 | mean accuracy = 0.9416
[valid] Epoch 5 mean loss = 0.2599 | mean accuracy = 0.9181

Epoch 6/10
[train] Epoch 6 mean loss = 0.0988 | mean accuracy = 0.9697
[valid] Epoch 6 mean loss = 0.2456 | mean accuracy = 0.9240

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 20:29:17,396] Trial 12 finished with value: 0.3921160674562641 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0004520724638002848, 'batch_size': 64}. Best is trial 11 with value: 0.26232793138307686.


[valid] Epoch 10 mean loss = 0.3834 | mean accuracy = 0.9304

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0001847078780156269, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 1.2486 | mean accuracy = 0.7143
[valid] Epoch 1 mean loss = 0.9922 | mean accuracy = 0.7724

Epoch 2/10
[train] Epoch 2 mean loss = 0.7919 | mean accuracy = 0.7760
[valid] Epoch 2 mean loss = 0.7203 | mean accuracy = 0.7946

Epoch 3/10
[train] Epoch 3 mean loss = 0.5853 | mean accuracy = 0.8252
[valid] Epoch 3 mean loss = 0.5472 | mean accuracy = 0.8361

Epoch 4/10
[train] Epoch 4 mean loss = 0.4430 | mean accuracy = 0.8662
[valid] Epoch 4 mean loss = 0.4386 | mean accuracy = 0.8642

Epoch 5/10
[train] Epoch 5 mean loss = 0.3259 | mean accuracy = 0.8997
[valid] Epoch 5 mean loss = 0.3489 | mean accuracy = 0.8861

Epoch 6/10
[train] Epoch 6 mean loss = 0.2295 | mean accuracy = 0.9290
[valid] Epoch 6 mean loss = 0.2825 | mean accuracy = 0.9056

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 20:33:43,436] Trial 13 finished with value: 0.2262512398701088 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0001847078780156269, 'batch_size': 64}. Best is trial 13 with value: 0.2262512398701088.


[valid] Epoch 10 mean loss = 0.2172 | mean accuracy = 0.9358

Testing configuration: embedding_size=64, hidden_size=256, learning_rate=0.00010604025377179749, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 1.2693 | mean accuracy = 0.7355
[valid] Epoch 1 mean loss = 1.0352 | mean accuracy = 0.7696

Epoch 2/10
[train] Epoch 2 mean loss = 0.8221 | mean accuracy = 0.7747
[valid] Epoch 2 mean loss = 0.7666 | mean accuracy = 0.7984

Epoch 3/10
[train] Epoch 3 mean loss = 0.6246 | mean accuracy = 0.8187
[valid] Epoch 3 mean loss = 0.6159 | mean accuracy = 0.8189

Epoch 4/10
[train] Epoch 4 mean loss = 0.4989 | mean accuracy = 0.8528
[valid] Epoch 4 mean loss = 0.5325 | mean accuracy = 0.8395

Epoch 5/10
[train] Epoch 5 mean loss = 0.3834 | mean accuracy = 0.8841
[valid] Epoch 5 mean loss = 0.4402 | mean accuracy = 0.8630

Epoch 6/10
[train] Epoch 6 mean loss = 0.2909 | mean accuracy = 0.9099
[valid] Epoch 6 mean loss = 0.3743 | mean accuracy = 0.8834

Epoch 7/10
[train] Epoch 7 mean lo

[I 2024-11-20 20:39:14,181] Trial 14 finished with value: 0.29498225684259455 and parameters: {'embedding_size': 64, 'hidden_size': 256, 'learning_rate': 0.00010604025377179749, 'batch_size': 64}. Best is trial 13 with value: 0.2262512398701088.


[valid] Epoch 10 mean loss = 0.2879 | mean accuracy = 0.9193

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.00021755877534950923, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 1.1855 | mean accuracy = 0.7313
[valid] Epoch 1 mean loss = 0.9664 | mean accuracy = 0.7712

Epoch 2/10
[train] Epoch 2 mean loss = 0.7370 | mean accuracy = 0.7941
[valid] Epoch 2 mean loss = 0.6937 | mean accuracy = 0.8101

Epoch 3/10
[train] Epoch 3 mean loss = 0.5350 | mean accuracy = 0.8440
[valid] Epoch 3 mean loss = 0.5499 | mean accuracy = 0.8379

Epoch 4/10
[train] Epoch 4 mean loss = 0.4068 | mean accuracy = 0.8762
[valid] Epoch 4 mean loss = 0.4490 | mean accuracy = 0.8657

Epoch 5/10
[train] Epoch 5 mean loss = 0.2975 | mean accuracy = 0.9075
[valid] Epoch 5 mean loss = 0.3660 | mean accuracy = 0.8751

Epoch 6/10
[train] Epoch 6 mean loss = 0.2012 | mean accuracy = 0.9372
[valid] Epoch 6 mean loss = 0.3105 | mean accuracy = 0.8906

Epoch 7/10
[train] Epoch 7 mean lo

[I 2024-11-20 20:43:40,715] Trial 15 finished with value: 0.2946326095683902 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.00021755877534950923, 'batch_size': 64}. Best is trial 13 with value: 0.2262512398701088.


[valid] Epoch 10 mean loss = 0.2938 | mean accuracy = 0.9080

Testing configuration: embedding_size=32, hidden_size=128, learning_rate=0.0002080928795447582, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 1.2306 | mean accuracy = 0.6846
[valid] Epoch 1 mean loss = 1.0164 | mean accuracy = 0.7693

Epoch 2/10
[train] Epoch 2 mean loss = 0.8195 | mean accuracy = 0.7735
[valid] Epoch 2 mean loss = 0.7823 | mean accuracy = 0.7791

Epoch 3/10
[train] Epoch 3 mean loss = 0.6459 | mean accuracy = 0.8068
[valid] Epoch 3 mean loss = 0.6350 | mean accuracy = 0.8148

Epoch 4/10
[train] Epoch 4 mean loss = 0.5155 | mean accuracy = 0.8442
[valid] Epoch 4 mean loss = 0.5447 | mean accuracy = 0.8373

Epoch 5/10
[train] Epoch 5 mean loss = 0.3970 | mean accuracy = 0.8771
[valid] Epoch 5 mean loss = 0.4444 | mean accuracy = 0.8621

Epoch 6/10
[train] Epoch 6 mean loss = 0.3000 | mean accuracy = 0.9058
[valid] Epoch 6 mean loss = 0.3652 | mean accuracy = 0.8873

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 20:48:08,625] Trial 16 finished with value: 0.27874122501588333 and parameters: {'embedding_size': 32, 'hidden_size': 128, 'learning_rate': 0.0002080928795447582, 'batch_size': 64}. Best is trial 13 with value: 0.2262512398701088.


[valid] Epoch 10 mean loss = 0.2687 | mean accuracy = 0.9145

Testing configuration: embedding_size=64, hidden_size=128, learning_rate=0.0007070995565111161, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 0.9349 | mean accuracy = 0.7602
[valid] Epoch 1 mean loss = 0.7191 | mean accuracy = 0.8010

Epoch 2/10
[train] Epoch 2 mean loss = 0.5015 | mean accuracy = 0.8483
[valid] Epoch 2 mean loss = 0.4650 | mean accuracy = 0.8537

Epoch 3/10
[train] Epoch 3 mean loss = 0.3031 | mean accuracy = 0.9059
[valid] Epoch 3 mean loss = 0.3429 | mean accuracy = 0.8898

Epoch 4/10
[train] Epoch 4 mean loss = 0.1907 | mean accuracy = 0.9410
[valid] Epoch 4 mean loss = 0.2719 | mean accuracy = 0.9137

Epoch 5/10
[train] Epoch 5 mean loss = 0.1239 | mean accuracy = 0.9617
[valid] Epoch 5 mean loss = 0.2484 | mean accuracy = 0.9254

Epoch 6/10
[train] Epoch 6 mean loss = 0.0552 | mean accuracy = 0.9836
[valid] Epoch 6 mean loss = 0.2294 | mean accuracy = 0.9380

Epoch 7/10
[train] Epoch 7 mean los

[I 2024-11-20 20:52:49,400] Trial 17 finished with value: 0.3016170789213741 and parameters: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0007070995565111161, 'batch_size': 64}. Best is trial 13 with value: 0.2262512398701088.


[valid] Epoch 10 mean loss = 0.2882 | mean accuracy = 0.9481

Testing configuration: embedding_size=64, hidden_size=256, learning_rate=0.00020082829074530813, batch_size=64

Epoch 1/10
[train] Epoch 1 mean loss = 1.1038 | mean accuracy = 0.7334
[valid] Epoch 1 mean loss = 0.9120 | mean accuracy = 0.7764

Epoch 2/10
[train] Epoch 2 mean loss = 0.6856 | mean accuracy = 0.8087
[valid] Epoch 2 mean loss = 0.6244 | mean accuracy = 0.8172

Epoch 3/10
[train] Epoch 3 mean loss = 0.4899 | mean accuracy = 0.8545
[valid] Epoch 3 mean loss = 0.5036 | mean accuracy = 0.8461

Epoch 4/10
[train] Epoch 4 mean loss = 0.3790 | mean accuracy = 0.8859
[valid] Epoch 4 mean loss = 0.4234 | mean accuracy = 0.8693

Epoch 5/10
[train] Epoch 5 mean loss = 0.2862 | mean accuracy = 0.9113
[valid] Epoch 5 mean loss = 0.3625 | mean accuracy = 0.8760

Epoch 6/10
[train] Epoch 6 mean loss = 0.1877 | mean accuracy = 0.9415
[valid] Epoch 6 mean loss = 0.3014 | mean accuracy = 0.8950

Epoch 7/10
[train] Epoch 7 mean lo

[I 2024-11-20 20:58:53,082] Trial 18 finished with value: 0.3206683176697469 and parameters: {'embedding_size': 64, 'hidden_size': 256, 'learning_rate': 0.00020082829074530813, 'batch_size': 64}. Best is trial 13 with value: 0.2262512398701088.


[valid] Epoch 10 mean loss = 0.3125 | mean accuracy = 0.9090

Testing configuration: embedding_size=32, hidden_size=64, learning_rate=0.0003426719736850518, batch_size=32

Epoch 1/10
[train] Epoch 1 mean loss = 0.9998 | mean accuracy = 0.7547
[valid] Epoch 1 mean loss = 0.8542 | mean accuracy = 0.7741

Epoch 2/10
[train] Epoch 2 mean loss = 0.7207 | mean accuracy = 0.7846
[valid] Epoch 2 mean loss = 0.7050 | mean accuracy = 0.7970

Epoch 3/10
[train] Epoch 3 mean loss = 0.5636 | mean accuracy = 0.8268
[valid] Epoch 3 mean loss = 0.6065 | mean accuracy = 0.8273

Epoch 4/10
[train] Epoch 4 mean loss = 0.4325 | mean accuracy = 0.8667
[valid] Epoch 4 mean loss = 0.5376 | mean accuracy = 0.8520

Epoch 5/10
[train] Epoch 5 mean loss = 0.3163 | mean accuracy = 0.9010
[valid] Epoch 5 mean loss = 0.4845 | mean accuracy = 0.8729

Epoch 6/10
[train] Epoch 6 mean loss = 0.2285 | mean accuracy = 0.9274
[valid] Epoch 6 mean loss = 0.4476 | mean accuracy = 0.8947

Epoch 7/10
[train] Epoch 7 mean loss

[I 2024-11-20 21:03:17,825] Trial 19 finished with value: 0.4403449323247461 and parameters: {'embedding_size': 32, 'hidden_size': 64, 'learning_rate': 0.0003426719736850518, 'batch_size': 32}. Best is trial 13 with value: 0.2262512398701088.


[valid] Epoch 10 mean loss = 0.4331 | mean accuracy = 0.9226

Best configuration:
{'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0001847078780156269, 'batch_size': 64}
Best validation loss: 0.2262512398701088

Total time taken: 6121.59 seconds


# Improvements

To improve the NER Tagger we firsly need to modify `vocabulary` function by adding there POS tag and Characters mapping, to add `read_conll_pos` function for POS tags, to add new function for caharcters extraction, a new `datagenerator` with pos_sym2idx, pos_idx2sym, char_sym2idx, char_idx2sym.

In [ ]:
# NEW VOCABULARY with pos_tag and char mapping
def vocabulary(filename, vocab_type, padding='<pad>', unknown='<unk>'):
    # vocab_type tells the type of vocabulary to create: 'input', 'output', 'pos', 'char'
    # the optional flags indicate that a padding and an unknown token have to be added to the vocabulary
    # if their value is not None

    idx2sym = {}
    sym2idx = {}

    cur_idx = 0
    # Add pad and unk tokens to the vocab if applicable
    if padding:
      idx2sym[cur_idx] = padding
      sym2idx[padding] = cur_idx
      cur_idx += 1
    if unknown:
      idx2sym[cur_idx] = unknown
      sym2idx[unknown] = cur_idx
      cur_idx += 1

    with open(filename, 'r', encoding='utf-8') as f:
      for line in f:
        line = line.strip()

        if not line or line.startswith('-DOCSTART-'):
          continue  # Skip empty lines and metadata

        parts = line.split()
        if vocab_type == 'input':
          token = parts[0]  # Token
          if token not in sym2idx:
            idx2sym[cur_idx] = token
            sym2idx[token] = cur_idx
            cur_idx += 1

        elif vocab_type == 'output':
          tag = parts[-1]  # NER tag
          if tag not in sym2idx:
            idx2sym[cur_idx] = tag
            sym2idx[tag] = cur_idx
            cur_idx += 1

        elif vocab_type == 'pos':
          pos_tag = parts[1]  # POS tag
          if pos_tag not in sym2idx:
            idx2sym[cur_idx] = pos_tag
            sym2idx[pos_tag] = cur_idx
            cur_idx += 1

        elif vocab_type == 'char':
          token = parts[0]  # Token
          for char in token:  # Extract all characters from the token
            if char not in sym2idx:
              idx2sym[cur_idx] = char
              sym2idx[char] = cur_idx
              cur_idx += 1

    return idx2sym, sym2idx

In [ ]:
# Test
# Tokens
input_idx2sym, input_sym2idx = vocabulary('eng.train', vocab_type='input')
print(input_idx2sym)
print(input_sym2idx)
# NER tags
output_idx2sym, output_sym2idx = vocabulary('eng.train', vocab_type='output')
print(output_idx2sym)
print(output_sym2idx)
# POS tags
pos_idx2sym, pos_sym2idx = vocabulary('eng.train', vocab_type='pos')
print(pos_idx2sym)
print(pos_sym2idx)
# Characters
char_idx2sym, char_sym2idx = vocabulary('eng.train', vocab_type='char')
print(char_idx2sym)
print(char_sym2idx)

{0: '<pad>', 1: '<unk>', 2: 'EU', 3: 'rejects', 4: 'German', 5: 'call', 6: 'to', 7: 'boycott', 8: 'British', 9: 'lamb', 10: '.', 11: 'Peter', 12: 'Blackburn', 13: 'BRUSSELS', 14: '1996-08-22', 15: 'The', 16: 'European', 17: 'Commission', 18: 'said', 19: 'on', 20: 'Thursday', 21: 'it', 22: 'disagreed', 23: 'with', 24: 'advice', 25: 'consumers', 26: 'shun', 27: 'until', 28: 'scientists', 29: 'determine', 30: 'whether', 31: 'mad', 32: 'cow', 33: 'disease', 34: 'can', 35: 'be', 36: 'transmitted', 37: 'sheep', 38: 'Germany', 39: "'s", 40: 'representative', 41: 'the', 42: 'Union', 43: 'veterinary', 44: 'committee', 45: 'Werner', 46: 'Zwingmann', 47: 'Wednesday', 48: 'should', 49: 'buy', 50: 'sheepmeat', 51: 'from', 52: 'countries', 53: 'other', 54: 'than', 55: 'Britain', 56: 'scientific', 57: 'was', 58: 'clearer', 59: '"', 60: 'We', 61: 'do', 62: "n't", 63: 'support', 64: 'any', 65: 'such', 66: 'recommendation', 67: 'because', 68: 'we', 69: 'see', 70: 'grounds', 71: 'for', 72: ',', 73: 'chie

In [ ]:
# ADD read_conll function for POS tags
def read_conll_pos(conllfilename):
    """
    Reads a CONLL 2003 file and returns a list of sentences.
    A sentence is a list of strings (POS tags)
    """
    #TODO
    sentences = []
    cur_sent = []

    with open(conllfilename, 'r') as f:
      for line in f:
        line = line.strip()
        if not line:
          if cur_sent:
            sentences.append(cur_sent)
            cur_sent = []
        elif not line.startswith('-DOCSTART-'):
        #else:
          parts = line.split()
          cur_sent.append(parts[1])   # POS tag

    if cur_sent:
      sentences.append(cur_sent)

    return sentences

In [ ]:
# Test
pos_tags = read_conll_pos('eng.train')
print(pos_tags)

[['NNP', 'VBZ', 'JJ', 'NN', 'TO', 'VB', 'JJ', 'NN', '.'], ['NNP', 'NNP'], ['NNP', 'CD'], ['DT', 'NNP', 'NNP', 'VBD', 'IN', 'NNP', 'PRP', 'VBD', 'IN', 'JJ', 'NN', 'TO', 'NNS', 'TO', 'VB', 'JJ', 'NN', 'IN', 'NNS', 'VBP', 'IN', 'JJ', 'NN', 'NN', 'MD', 'VB', 'VBN', 'TO', 'NN', '.'], ['NNP', 'POS', 'NN', 'TO', 'DT', 'NNP', 'NNP', 'POS', 'JJ', 'NN', 'NNP', 'NNP', 'VBD', 'IN', 'NNP', 'NNS', 'MD', 'VB', 'NN', 'IN', 'NNS', 'JJ', 'IN', 'NNP', 'IN', 'DT', 'JJ', 'NN', 'VBD', 'JJR', '.'], ['"', 'PRP', 'VBP', 'RB', 'VB', 'DT', 'JJ', 'NN', 'IN', 'PRP', 'VBP', 'RB', 'VB', 'DT', 'NNS', 'IN', 'PRP', ',', '"', 'DT', 'NNP', 'POS', 'JJ', 'NN', 'NNP', 'NNP', 'FW', 'NNP', 'VBD', 'DT', 'NN', 'NN', '.'], ['PRP', 'VBD', 'JJ', 'JJ', 'NN', 'VBD', 'VBN', 'CC', 'IN', 'PRP', 'VBD', 'VBN', 'IN', 'NN', 'VBD', 'VBN', 'PRP', 'MD', 'VB', 'VBN', 'IN', 'DT', 'NNP', 'NNP', '.'], ['PRP', 'VBD', 'DT', 'NN', 'JJ', 'NN', 'IN', 'NNP', 'NNP', 'NNP', 'NNP', 'NNP', 'TO', 'VB', 'NN', 'NNS', ',', 'NNS', 'CC', 'JJ', 'NNS', 'IN', 'DT',

In [ ]:
# NEW function to extract characters
def extract_char_sequences(tokens):
    char_sequences = []
    for sentence in tokens:
      char_sequences.append([[char for char in token] for token in sentence])
    return char_sequences

In [ ]:
# Test on first sentence Pad sequences, Encode sequences and Decode sequences
sample_tokens = tokens[0]
sample_tags = tags[0]
sample_pos = pos_tags[0]
sample_chars = char_sequences[0]
print(sample_tokens)
print(sample_tags)
print(sample_pos)
print(sample_chars)
print()

# Pad sequences
padded_tokens = pad_sequence(sample_tokens, pad_size=12, pad_token='<pad>')
padded_tags = pad_sequence(sample_tags, pad_size=12, pad_token='<pad>')
padded_pos = pad_sequence(sample_pos, pad_size=12, pad_token='<pad>')
padded_chars = [pad_sequence(char_seq, pad_size=15, pad_token='<pad>') for char_seq in sample_chars]
print(padded_tokens)
print(padded_tags)
print(padded_pos)
print(padded_chars)
print()

# Encode sequences
encoded_tokens = code_sequence(padded_tokens, input_sym2idx, unk_token='<unk>')
encoded_tags = code_sequence(padded_tags, output_sym2idx, unk_token='<unk>')
encoded_pos = code_sequence(padded_pos, pos_sym2idx, unk_token='<unk>')
encoded_chars = [code_sequence(char_seq, char_sym2idx, unk_token='<unk>') for char_seq in padded_chars]
print(encoded_tokens)
print(encoded_tags)
print(encoded_pos)
print(encoded_chars)
print()

# Decode sequences
decoded_tokens = decode_sequence(encoded_tokens, input_idx2sym)
decoded_tags = decode_sequence(encoded_tags, output_idx2sym)
decoded_pos = decode_sequence(encoded_pos, pos_idx2sym)
decoded_chars = [decode_sequence(char_seq, char_idx2sym) for char_seq in encoded_chars]
print(decoded_tokens)
print(decoded_tags)
print(decoded_pos)
print(decoded_chars)

['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.']
['I-ORG', 'O', 'I-MISC', 'O', 'O', 'O', 'I-MISC', 'O', 'O']
['NNP', 'VBZ', 'JJ', 'NN', 'TO', 'VB', 'JJ', 'NN', '.']
[['E', 'U'], ['r', 'e', 'j', 'e', 'c', 't', 's'], ['G', 'e', 'r', 'm', 'a', 'n'], ['c', 'a', 'l', 'l'], ['t', 'o'], ['b', 'o', 'y', 'c', 'o', 't', 't'], ['B', 'r', 'i', 't', 'i', 's', 'h'], ['l', 'a', 'm', 'b'], ['.']]

['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.', '<pad>', '<pad>', '<pad>']
['I-ORG', 'O', 'I-MISC', 'O', 'O', 'O', 'I-MISC', 'O', 'O', '<pad>', '<pad>', '<pad>']
['NNP', 'VBZ', 'JJ', 'NN', 'TO', 'VB', 'JJ', 'NN', '.', '<pad>', '<pad>', '<pad>']
[['E', 'U', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>'], ['r', 'e', 'j', 'e', 'c', 't', 's', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>'], ['G', 'e', 'r', 'm', 'a', 'n', '<pad>', '<pad>', '<pad>', '<pad>', '<pad>'

In [ ]:
# NEW datagenerator with pos_sym2idx, pos_idx2sym, char_sym2idx, char_idx2sym
import torch
import torch.nn as nn
from random import shuffle

class DataGenerator:

        def __init__(self,conllfilename, parentgenerator = None, pad_token='<pad>',unk_token='<unk>'):

              if parentgenerator is not None:
                  self.pad_token = parentgenerator.pad_token
                  self.unk_token = parentgenerator.unk_token
                  self.input_sym2idx = parentgenerator.input_sym2idx
                  self.input_idx2sym = parentgenerator.input_idx2sym
                  self.output_sym2idx = parentgenerator.output_sym2idx
                  self.output_idx2sym = parentgenerator.output_idx2sym
                  self.pos_sym2idx = parentgenerator.pos_sym2idx
                  self.pos_idx2sym = parentgenerator.pos_idx2sym
                  self.char_sym2idx = parentgenerator.char_sym2idx
                  self.char_idx2sym = parentgenerator.char_idx2sym
              else:                           # Creates new encodings
                  self.pad_token = pad_token
                  self.unk_token = unk_token
                  #TODO : Create 8 encoding maps from datafile
                  self.input_idx2sym, self.input_sym2idx = vocabulary(conllfilename, vocab_type='input', padding=pad_token, unknown=unk_token)
                  self.output_idx2sym, self.output_sym2idx = vocabulary(conllfilename, vocab_type='output', padding=pad_token, unknown=unk_token)
                  self.pos_idx2sym, self.pos_sym2idx = vocabulary(conllfilename, vocab_type='pos', padding=pad_token, unknown=unk_token)
                  self.char_idx2sym, self.char_sym2idx = vocabulary(conllfilename, vocab_type='char', padding=pad_token, unknown=unk_token)

              #TODO : store the conll dataset with sentence structure (a list of lists of strings) in the following fields
              self.Xtokens = read_conll_tokens(conllfilename)
              self.Ytokens = read_conll_tags(conllfilename)
              self.Xpos = read_conll_pos(conllfilename)
              self.Xchars = extract_char_sequences(self.Xtokens)

        def generate_batches(self,batch_size):

              # Generator function yielding one batch after another. Batches are lists of lists

              assert(len(self.Xtokens) == len(self.Ytokens) == len(self.Xpos) == len(self.Xchars))

              N = len(self.Xtokens)
              idxes = list(range(N))

              # Data ordering
              shuffle(idxes)
              idxes.sort(key=lambda idx: len(self.Xtokens[idx]))

              # batch generation
              bstart = 0
              while bstart < N:
                 bend        = min(bstart+batch_size,N)
                 batch_idxes = idxes[bstart:bend]
                 batch_len   = max(len(self.Xtokens[idx]) for idx in batch_idxes)

                 # Pad sequences (tokens, NER tags, POS tags)
                 seqX = [pad_sequence(self.Xtokens[idx],batch_len,self.pad_token) for idx in batch_idxes]
                 seqY = [pad_sequence(self.Ytokens[idx],batch_len,self.pad_token) for idx in batch_idxes]
                 seqPOS = [pad_sequence(self.Xpos[idx], batch_len, self.pad_token) for idx in batch_idxes]

                 # Encode sequences (tokens, NER tags, POS tags)
                 seqX = [code_sequence(seq,self.input_sym2idx,self.unk_token) for seq in seqX]
                 seqY = [code_sequence(seq,self.output_sym2idx) for seq in seqY]
                 seqPOS = [code_sequence(seq, self.pos_sym2idx, self.unk_token) for seq in seqPOS]
                 #print(seqX, seqY, seqPOS)

                 # Character sequences
                 char_batch = []
                 for i in batch_idxes:
                   char_sequences = self.Xchars[i]  # list of character sequences for each token in a sentence
                   padded_char_sequences = [pad_sequence(char_seq, pad_size=15, pad_token=self.pad_token) for char_seq in char_sequences]
                   encoded_char_sequences = [code_sequence(char_seq, self.char_sym2idx, self.unk_token) for char_seq in padded_char_sequences]
                   char_batch.append(encoded_char_sequences)

                 # Pad character sequences for each word in the sentence
                 max_sent_len = batch_len
                 max_word_len = 15
                 # Ensure all sentences in the batch have the same length for character sequences
                 seqChar = [pad_sequence(char_seq, max_sent_len, [self.char_sym2idx[self.pad_token]] * max_word_len) for char_seq in char_batch]

                 assert(len(seqX) == len(seqY) == len(seqPOS) == len(seqChar))
                 yield (seqX,seqY,seqPOS, seqChar)
                 bstart += batch_size

# Improved NER Tagger

In [ ]:
# NEW NER Tagger
import torch.optim as optim

import torch
import torch.nn as nn
import torch.optim as optim
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence

class NERtagger(nn.Module):

    def __init__(self, traingenerator, embedding_size, hidden_size, pos_embedding_size, char_embedding_size, cnn_filters, device='cuda'):
        super(NERtagger, self).__init__()
        self.embedding_size = embedding_size
        self.hidden_size = hidden_size
        self.pos_embedding_size = pos_embedding_size
        self.char_embedding_size = char_embedding_size
        self.cnn_filters = cnn_filters
        self.allocate_params(traingenerator, device)

    def allocate_params(self, datagenerator, device):
        # Vocabulary and output sizes
        vocab_size = len(datagenerator.input_sym2idx)
        num_classes = len(datagenerator.output_sym2idx)
        pos_vocab_size = len(datagenerator.pos_sym2idx)
        char_vocab_size = len(datagenerator.char_sym2idx)

        # Token embeddings
        self.embedding = nn.Embedding(vocab_size, self.embedding_size, padding_idx=datagenerator.input_sym2idx[datagenerator.pad_token]).to(device)

        # POS embeddings
        self.pos_embedding = nn.Embedding(pos_vocab_size, self.pos_embedding_size, padding_idx=datagenerator.pos_sym2idx[datagenerator.pad_token]).to(device)

        # Character embeddings and CNN (1D convolution)
        self.char_embedding = nn.Embedding(char_vocab_size, self.char_embedding_size, padding_idx=datagenerator.char_sym2idx[datagenerator.pad_token]).to(device)
        self.char_cnn = nn.Conv1d(self.char_embedding_size, self.cnn_filters, kernel_size=3, padding=1).to(device)

        # Bi-LSTM
        self.lstm = nn.LSTM(
            input_size=self.embedding_size + self.pos_embedding_size + self.cnn_filters,
            hidden_size=self.hidden_size,
            num_layers=1,
            batch_first=True,
            bidirectional=True
        ).to(device)

        # Attention layer
        self.attention = nn.Linear(self.hidden_size * 2, 1, bias=False).to(device)

        # Output layer
        self.linear = nn.Linear(self.hidden_size * 2, num_classes).to(device)

    def compute_attention(self, lstm_out):
        # lstm_out: (batch_size, seq_len, hidden_size * 2)
        attention_scores = self.attention(lstm_out)  # Shape: (batch_size, seq_len, 1)
        attention_weights = torch.softmax(attention_scores, dim=1)  # Shape: (batch_size, seq_len, 1)
        context = torch.sum(attention_weights * lstm_out, dim=1)  # Shape: (batch_size, hidden_size * 2)
        return context

    def forward(self, Xinput, Xpos, Xchar):
        # Embedding layers
        token_embeddings = self.embedding(Xinput)  # Shape: (batch_size, seq_len, embedding_size)
        pos_embeddings = self.pos_embedding(Xpos)  # Shape: (batch_size, seq_len, pos_embedding_size)

        # Character embeddings + CNN
        batch_size, seq_len, word_len = Xchar.shape
        Xchar_flat = Xchar.view(-1, word_len)  # Shape: (batch_size * seq_len, word_len)
        char_embeddings = self.char_embedding(Xchar_flat).permute(0, 2, 1)  # Shape: (batch_size * seq_len, char_embedding_size, word_len)
        char_features = torch.relu(self.char_cnn(char_embeddings))  # Shape: (batch_size * seq_len, cnn_filters, word_len)
        # max pooling operation
        char_features = torch.max(char_features, dim=2).values  # Shape: (batch_size * seq_len, cnn_filters)
        char_features = char_features.view(batch_size, seq_len, -1)  # Shape: (batch_size, seq_len, cnn_filters)

        # Concatenate features
        embeddings = torch.cat([token_embeddings, pos_embeddings, char_features], dim=2)  # Shape: (batch_size, seq_len, embedding_size + pos_embedding_size + cnn_filters)

        # Bi-LSTM
        lstm_out, _ = self.lstm(embeddings)  # Shape: (batch_size, seq_len, hidden_size * 2)

        # Attention
        context = self.compute_attention(lstm_out)  # Shape: (batch_size, hidden_size * 2)

        # Output Layer
        logits = self.linear(lstm_out)  # Shape: (batch_size, seq_len, num_classes)
        return logits

    def train_model(self, traingenerator, validgenerator, epochs, batch_size, device='cuda', learning_rate=0.001):
        self.minloss = float('inf')
        self.to(device)
        optimizer = optim.Adam(self.parameters(), lr=learning_rate)
        pad_index = traingenerator.output_sym2idx[traingenerator.pad_token]
        loss_fnc = nn.CrossEntropyLoss(ignore_index=pad_index)

        for epoch in range(1, epochs + 1):
            print(f"\nEpoch {epoch}/{epochs}")
            self.train()
            batch_losses, batch_accuracies = [], []

            for seqX, seqY, seqPOS, seqChar in traingenerator.generate_batches(batch_size):
                X = torch.LongTensor(seqX).to(device)
                Y = torch.LongTensor(seqY).to(device)
                POS = torch.LongTensor(seqPOS).to(device)
                CHAR = torch.LongTensor(seqChar).to(device)

                Yhat = self.forward(X, POS, CHAR)

                # Flatten for loss computation
                batch_size, seq_len = Y.shape
                Yhat = Yhat.view(batch_size * seq_len, -1)
                Y = Y.view(batch_size * seq_len)

                # Compute loss
                loss = loss_fnc(Yhat, Y)
                batch_losses.append(loss.item())

                # Backpropagation
                optimizer.zero_grad()
                loss.backward()
                optimizer.step()

                # Compute accuracy
                mask = (Y != pad_index)
                Yargmax = torch.argmax(Yhat, dim=1)
                correct = torch.sum((Yargmax == Y) * mask)
                total = torch.sum(mask)
                batch_accuracies.append(float(correct) / float(total))

            # Epoch summary
            train_loss = sum(batch_losses) / len(batch_losses)
            train_accuracy = sum(batch_accuracies) / len(batch_accuracies)
            print(f"[train] Epoch {epoch} mean loss = {train_loss:.4f} | mean accuracy = {train_accuracy:.4f}")

            valid_loss, valid_accuracy = self.validate(validgenerator, batch_size, device, save_min_model=True)
            print(f"[valid] Epoch {epoch} mean loss = {valid_loss:.4f} | mean accuracy = {valid_accuracy:.4f}")

    def validate(self, datagenerator, batch_size, device='cuda', save_min_model=False):
        batch_losses, batch_accuracies = [], []
        pad_index = datagenerator.output_sym2idx[datagenerator.pad_token]
        loss_fnc = nn.CrossEntropyLoss(ignore_index=pad_index)

        for seqX, seqY, seqPOS, seqChar in datagenerator.generate_batches(batch_size):
            with torch.no_grad():
                X = torch.LongTensor(seqX).to(device)
                Y = torch.LongTensor(seqY).to(device)
                POS = torch.LongTensor(seqPOS).to(device)
                CHAR = torch.LongTensor(seqChar).to(device)

                Yhat = self.forward(X, POS, CHAR)

                # Flatten for loss computation
                batch_size, seq_len = Y.shape
                Yhat = Yhat.view(batch_size * seq_len, -1)
                Y = Y.view(batch_size * seq_len)

                # Compute loss
                loss = loss_fnc(Yhat, Y)
                batch_losses.append(loss.item())

                # Compute accuracy
                mask = (Y != pad_index)
                Yargmax = torch.argmax(Yhat, dim=1)
                correct = torch.sum((Yargmax == Y) * mask)
                total = torch.sum(mask)
                batch_accuracies.append(float(correct) / float(total))

        # Validation summary
        valid_loss = sum(batch_losses) / len(batch_losses)
        valid_accuracy = sum(batch_accuracies) / len(batch_accuracies)

        if save_min_model and valid_loss < self.minloss:
            self.minloss = valid_loss
            torch.save(self.state_dict(), 'tagger_params.pt')

        return valid_loss, valid_accuracy

In [ ]:
# Train on the full dataset
trainset_i = DataGenerator('eng.train')
validset_i = DataGenerator('eng.testa',parentgenerator = trainset_i)
tagger_i = NERtagger(trainset_i,embedding_size=64,hidden_size=128,pos_embedding_size=32,char_embedding_size=25,cnn_filters=30,device='cuda')
tagger_i.train_model(traingenerator=trainset_i,validgenerator=validset_i,epochs=13,batch_size=64,device='cuda',learning_rate=0.0001847)


Epoch 1/13
[train] Epoch 1 mean loss = 0.8721 | mean accuracy = 0.7670
[valid] Epoch 1 mean loss = 0.5775 | mean accuracy = 0.8145

Epoch 2/13
[train] Epoch 2 mean loss = 0.4182 | mean accuracy = 0.8625
[valid] Epoch 2 mean loss = 0.4084 | mean accuracy = 0.8557

Epoch 3/13
[train] Epoch 3 mean loss = 0.2785 | mean accuracy = 0.9116
[valid] Epoch 3 mean loss = 0.2738 | mean accuracy = 0.9194

Epoch 4/13
[train] Epoch 4 mean loss = 0.2009 | mean accuracy = 0.9389
[valid] Epoch 4 mean loss = 0.2088 | mean accuracy = 0.9377

Epoch 5/13
[train] Epoch 5 mean loss = 0.1463 | mean accuracy = 0.9549
[valid] Epoch 5 mean loss = 0.1897 | mean accuracy = 0.9424

Epoch 6/13
[train] Epoch 6 mean loss = 0.1001 | mean accuracy = 0.9693
[valid] Epoch 6 mean loss = 0.1543 | mean accuracy = 0.9547

Epoch 7/13
[train] Epoch 7 mean loss = 0.0696 | mean accuracy = 0.9792
[valid] Epoch 7 mean loss = 0.1386 | mean accuracy = 0.9577

Epoch 8/13
[train] Epoch 8 mean loss = 0.0479 | mean accuracy = 0.9860
[val

# Conclusion

**Hyperparameters search**
We used the hyperparameter optimization framework Optuna and we searched for the best configuration of hyperparameters by minimizing the validation loss.

The best configuration found is: {'embedding_size': 64, 'hidden_size': 128, 'learning_rate': 0.0001847078780156269, 'batch_size': 64} with the best validation loss: 0.23.

The training results are as follows:\
The train loss: 0.0535\
Tthe train accuracy: 98.50%\
The validation results are as follows:\
The valid loss: 0.2172\
The valid accuracy: 93.58%

The model maintains a reasonably good balance between memorizing training data and generalizing to unseen validation data. The difference between the validation accuracy (93.58%) and the training accuracy (98.50%) is 5%, indicating minor overfitting but still reasonable generalizability.

**NER Tagger improvements**

1) **Attention layer.** We added an attention layer to enhance the model's focus on key parts of the input sequence. This layer takes the Bi-LSTM output and produces a single attention score for each token. The softmax function is used to normalize these scores into `attention_weights`. The `context` vector then is computed as a weighted sum of the Bi-LSTM outputs, providing a richer representation in the `forward()` method.

2) **Part-of-speech (POS) tags embeddings.** We incorporated part-of-speech tag embeddings as additional inputs. First a separate embedding layer is defined for the POS tags, then in the `forward()` method POS tags are embedded using the POS embedding layer, and finally POS embeddings are concatenated with the token embeddings and character-level features.

3) **Character-level embeddings with Convolutional module for Unknown words.** To handle unknown words, we used a convolutional neural network (CNN) over character embeddings. First an embedding for each character in the vocabulary is created, then it is passed through the convolutional neural network CNN that applies filters to extract meaningful subword caracter-level features. As a result the model can generate an embedding for an unknown word by combining the features learned from its characters.

**Results**
With the improved model the best validation loss was achieved at the Epoch 9: 0.1279.
The train accuracy achieved is 99.06%.
The valid accuracy achieved is 96.34%.
After improvements the valid accuracy increased by 2.76% from 93.58% to 96.34%. These results demonstrate that the implementation of the attention mechanism, of the POS tag embeddings, and of the convolutional word embedding module improved the NER Tagger model's ability to generalize and accurately recognize named entities, making it more efficient.


